---
output:
  knitrBootstrap::bootstrap_document:
    title: "Kobzik - Transgenerational epigenetic inheritance of asthma susceptibility"
    theme: readable
    highlight: zenburn
    theme.chooser: FALSE
    highlight.chooser: FALSE


---

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Kobzik - Transgenerational epigenetic inheritance of asthma susceptibility"
clientname=" "
clientemail=""
labPI=""
lablocation=""
analystname="Andreas Sjodin"
analystemail="sjodin@hsph.harvard.edu"

```


```{r knitrsetup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", bootstrap.show.code=FALSE, bootstrap.show.output=FALSE, bootstrap.thumbnail.size="col-md-10", cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='', tidy.opts=list(keep.blank.line=FALSE, width.cutoff=200), fig.width = 16, fig.height = 14)

```


---

# Kobzik methylation



Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Setup

## Libraries

R libraries needed to perform analysis

```{r setup}
library(gridExtra)
library(knitr)
library(ggplot2)
library(CHBUtils)
library(RColorBrewer)
library(pheatmap)
library(methylKit)
library(dplyr)
```

## Variables
Working directories, files and other variables necessary to the analysis.

```{r load}

## Setup Data and Results directory variables
if(file.exists("/Volumes/orch/group_dir2/kobzik-rbss/Kobzik-rbbs/")){
  baseDir="/Volumes/orch/group_dir2/kobzik-rbss/Kobzik-rbbs/"
    } else if (file.exists("/Users/johnhutchinson/Work/projects/kobzik-rrbs/")){
    baseDir="/Users/johnhutchinson/Work/projects/kobzik-rrbs/"
    } 
#/Users/andreassjodin/hbc-projects/orchestra/Kobzik-rbbs/data/bed

dataDir <- file.path(baseDir, "data/bed")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

lo.count.cutoff <- 10
q.cutoff <- 0.01
```

## Functions 

``` {r functions}
cor.meth <- function(object, method="spearman"){
  meth.mat = getData(object)[, object@numCs.index]
  (getData(object)[,object@numCs.index] + getData(object)[,object@numTs.index] )    
  names(meth.mat) = object@sample.ids
  cor.mat <- ( cor(meth.mat,method=method) )
  cor.mat
}

plot.targetanno.cpgi <- function(mydiff, cpg, id){
  diffCpGann=annotate.WithFeature.Flank(mydiff,cpg$CpGi,cpg$shores,feature.name="CpGi",flank.name="shores")
  #getFeatsWithTargetsStats(diffCpGann,percentage=TRUE)
  #plotTargetAnnotation(diffCpGann,col=c("green","gray","white"), main=paste(id, "- differential methylation annotation"))
  plotTargetAnnotation(diffCpGann,col="Set1", main=paste(id, "- differential methylation annotation"))
 }

plot.targetanno.pei <- function(mydiff, gene, id){
  diffAnn=annotate.WithGenicParts(mydiff,gene)
  #getTargetAnnotationStats(diffAnn,percentage=TRUE,precedence=TRUE)
  #props=getTargetAnnotationStats(diffAnn,percentage=TRUE,precedence=T)
  #names(props)=paste( paste(round(props),"%"),sep=" ")
  plotTargetAnnotation(diffAnn,precedence=TRUE, col="Set1", main=paste(id, "- differential methylation annotation"))
}

annodiff2 <- function(diff, idx, gene, cpg, all=F){
  if (all){
    idx <- 1:nrow(diff)
  }
  dselect <- select(diff, idx)
  anno <- annotate.WithGenicParts(dselect,gene)
  gene.info <- getMembers(anno)
  cpgi.info <- getMembers(annotate.WithFeature.Flank(dselect,cpg$CpGi,cpg$shores,feature.name="CpGi",flank.name="shores"))
  
  assoc <- getAssociationWithTSS(anno)
  
  if (all){
    out <- data.frame(getData(dselect), assoc[,-1])
  } else {
    out <- data.frame(getData(dselect)[,1:3], assoc[,-1])
  }
  symbol <- convertIDs(out$feature.name, "ENSEMBL", "SYMBOL", org.Mm.eg.db, "useFirst")
  entrez <- convertIDs(out$feature.name, "ENSEMBL", "ENTREZID", org.Mm.eg.db, "useFirst")
  description <- convertIDs(out$feature.name, "ENSEMBL", "GENENAME", org.Mm.eg.db, "useFirst")
  refseq <- convertIDs(out$feature.name, "ENSEMBL", "REFSEQ", org.Mm.eg.db, "useFirst")
  
  out <- data.frame(out, gene.info, cpgi.info, symbol=symbol, entrez=entrez, description=description, refseq=refseq)
  return(out)
}

annodiff <- function(diff, idx, gene, cpg, all=F){
  if (all){
    idx <- 1:nrow(diff)
  }
  dselect <- select(diff, idx)
  anno <- annotate.WithGenicParts(dselect,gene)
  gene.info <- getMembers(anno)
  cpgi.info <- getMembers(annotate.WithFeature.Flank(dselect,cpg$CpGi,cpg$shores,feature.name="CpGi",flank.name="shores"))
  
  assoc <- getAssociationWithTSS(anno)
  
  if (all){
    out <- data.frame(getData(dselect), assoc[,-1])
  } else {
    out <- data.frame(getData(dselect)[,1:3], assoc[,-1])
  }
  symbol <- convertIDs(out$feature.name, "REFSEQ", "SYMBOL", org.Mm.eg.db, "useFirst")
  entrez <- convertIDs(out$feature.name, "REFSEQ", "ENTREZID", org.Mm.eg.db, "useFirst")
  description <- convertIDs(out$feature.name, "REFSEQ", "GENENAME", org.Mm.eg.db, "useFirst")
  ensembl <- convertIDs(out$feature.name, "REFSEQ", "ENSEMBL", org.Mm.eg.db, "useFirst")
  
  out <- data.frame(out, gene.info, cpgi.info, symbol=symbol, entrez=entrez, description=description, ensembl=ensembl)
  return(out)
}


convertIDs <- function( ids, from, to, db, ifMultiple=c("putNA", "useFirst")) {
  stopifnot( inherits( db, "AnnotationDb" ) )
  ifMultiple <- match.arg( ifMultiple )
  suppressWarnings( selRes <- AnnotationDbi::select(
    db, keys=ids, keytype=from, columns=c(from,to) ) )
  if ( ifMultiple == "putNA" ) {
    duplicatedIds <- selRes[ duplicated( selRes[,1] ), 1 ]
    selRes <- selRes[ ! selRes[,1] %in% duplicatedIds, ]
  }
  return( selRes[ match( ids, selRes[,1] ), 2 ] )
}

manhattan.plot <- function(x, main="", col= c("red", "blue"), cutoff = 3, ymax=NULL) {
  dm <- getData(x)
  dm$chr <- gsub("chr", "", dm$chr)
  dm <- dm[dm$chr != "X" & dm$chr != "Y",]
  dm$chr <- as.numeric(dm$chr)
  dm <- dm[,c(1,2,6,7)]
  colnames(dm)<- c("CHR", "S", "Q", "MD")
  gr.dm <- transformDfToGr(dm, seqnames = "CHR", start = "S", width=1)
  gr.dm <- keepSeqlevels(gr.dm, as.character(1:19))
  data(ideoCyto, package = "biovizBase")
  seqlengths(gr.dm) <- as.numeric(seqlengths(ideoCyto$mm10)[1:19])
  gr.dm <- gr.dm[!is.na(gr.dm$Q)]
  ## transform pvalue
  values(gr.dm)$pvalue <- -log10(values(gr.dm)$Q)
  if (is.null(ymax) ) {
    ymax <- max(values(gr.dm)$pvalue)
  }  
  ylim <- range(0, ymax)
  values(gr.dm)$pvalue[values(gr.dm)$pvalue > ymax] <- Inf
  plotGrandLinear(gr.dm, aes(y = pvalue), color = col, cutoff = cutoff, cutoff.color = "blue", cutoff.size = 0.2, main=main, ylim=ylim)
}




```

# Preperations

## Import Data and Metadata

### Metadata

Code needed to load metadata

```{r metadataload, bootstrap.show.output=TRUE}
metadata <- read.table(file.path(metaDir, "meta.txt"),sep="\t",header=T)

metadata.out <- metadata[,c("samplename", "exposure", "type" ,"cohort", "generation")]
kable(metadata.out)
``` 

### Data

Code needed to load methylation data

```{r dataload}
setwd(dataDir)
filelist10 <- dir(pattern="cpg")
testdata <- rep(0, length(metadata$generation))
testdata[metadata$generation == "cntr"] <-1

sample.order <- order(as.numeric(unlist(lapply(strsplit(filelist10, "\\."), function(x) x[2]))))
filelist10 <- filelist10[sample.order]

meth.quants <- read(as.list(filelist10), sample.id=as.list(as.character(metadata$samplename)),  assembly="mm10", treatment=testdata, context="CpG" )




#batch 1
meth.quants.b1F1 <- reorganize(meth.quants ,sample.ids=c("15-b1-cntr", "16-b1-cntr", "17-b1-cntr", "1-b1-F1", "2-b1-F1", "3-b1-F1", "4-b1-F1"), treatment=c(0,0,0,1,1,1,1) )
meth.quants.b1F2 <- reorganize(meth.quants ,sample.ids=c( "15-b1-cntr", "16-b1-cntr", "17-b1-cntr",  "5-b1-F2", "6-b1-F2", "7-b1-F2", "8-b1-F2"), treatment=c(0,0,0,1,1,1,1) )
meth.quants.b1F3 <- reorganize(meth.quants ,sample.ids=c( "15-b1-cntr", "16-b1-cntr", "17-b1-cntr",  "10-b1-F3", "11-b1-F3", "12-b1-F3", "13-b1-F3"), treatment=c(0,0,0,1,1,1,1) )

#batch 2
meth.quants.b2F1 <- reorganize(meth.quants ,sample.ids=c("123-b2-cntr","124-b2-cntr", "125-b2-cntr", "126-b2-cntr", "60-b2-F1", "61-b2-F1", "62-b2-F1", "63-b2-F1"), treatment=c(0,0,0,0,1,1,1,1) )
meth.quants.b2F2 <- reorganize(meth.quants ,sample.ids=c("123-b2-cntr","124-b2-cntr", "125-b2-cntr", "126-b2-cntr", "90-b2-F2", "91-b2-F2", "92-b2-F2", "93-b2-F2"), treatment=c(0,0,0,0,1,1,1,1) )
meth.quants.b2F3 <- reorganize(meth.quants ,sample.ids=c("123-b2-cntr","124-b2-cntr", "125-b2-cntr", "126-b2-cntr", "119-b2-F3", "120-b2-F3", "121-b2-F3", "122-b2-F3"), treatment=c(0,0,0,0,1,1,1,1) )


```

# Analysis

## Descriptive statistics

### Percent methylation distribution in all samples

```{r methstats1}

for(i in 1:32) { 
    nam <- paste("p", i, sep = "")
    assign(nam, getMethylationStats(meth.quants[[i]], plot=TRUE, labels=FALSE))
}
```

__Batch 1__

```{r methstats.plot.b1, fig.width = 16, fig.height = 64}
grid.arrange(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,p13,p14,p15,p16 , ncol=2)
```

__Batch 2__

```{r methstats.plot.b2, fig.width = 16, fig.height = 64}
grid.arrange(p17,p18,p19,p20,p21,p22,p23,p24,p25,p26,p27,p28,p29,p30,p31,p32 , ncol=2)
```

### CpG coverage in all samples

```{r methstats2}
#lapply(meth.quants, function(n) getCoverageStats(n, plot=TRUE, labels=TRUE))

for(i in 1:32) { 
    nam <- paste("p", i, sep = "")
    assign(nam, getCoverageStats(meth.quants[[i]], plot=TRUE, labels=TRUE))
}
```

__Batch 1__

```{r methstats.CpGplot.b1, fig.width = 16, fig.height = 64}
grid.arrange(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,p13,p14,p15,p16 , ncol=2)
```

__Batch 2__

```{r methstats.CpGplot.b2, fig.width = 16, fig.height = 64}
grid.arrange(p17,p18,p19,p20,p21,p22,p23,p24,p25,p26,p27,p28,p29,p30,p31,p32 , ncol=2)
```


### Filter by coverage 

- removes bases with very high coverage (above 99.9th percentile) to adjust for potential PCR biases
- also remove bases with very low coverage (<10)

```{r methcoveragefilter}
meth.quants.filtered <- filterByCoverage(meth.quants, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)    # methylBase object with 799774 rows

#batch 1
meth.quants.b1F1.filtered <- filterByCoverage(meth.quants.b1F1, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)  
meth.quants.b1F2.filtered <- filterByCoverage(meth.quants.b1F2, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)  
meth.quants.b1F3.filtered <- filterByCoverage(meth.quants.b1F3, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)  

#batch 2
meth.quants.b2F1.filtered <- filterByCoverage(meth.quants.b2F1, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)  
meth.quants.b2F2.filtered <- filterByCoverage(meth.quants.b2F2, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)  
meth.quants.b2F3.filtered <- filterByCoverage(meth.quants.b2F3, lo.count = lo.count.cutoff , lo.perc = NULL, hi.count = NULL, hi.perc = 99.9)  
```

### Normalize 

Normalizes coverage values between samples using a scaling factor derived from differences between median of coverage distributions


```{r methnormalization}
meth.quants.filtered.norm <- normalizeCoverage(meth.quants.filtered )

#batch 1
meth.quants.b1F1.filtered.norm <- normalizeCoverage(meth.quants.b1F1.filtered )
meth.quants.b1F2.filtered.norm <- normalizeCoverage(meth.quants.b1F2.filtered )
meth.quants.b1F3.filtered.norm <- normalizeCoverage(meth.quants.b1F3.filtered )

#batch 2
meth.quants.b2F1.filtered.norm <- normalizeCoverage(meth.quants.b2F1.filtered )
meth.quants.b2F2.filtered.norm <- normalizeCoverage(meth.quants.b2F2.filtered )
meth.quants.b2F3.filtered.norm <- normalizeCoverage(meth.quants.b2F3.filtered )

```

### Merge samples

- only compare samples at sites where they all have coverage
- reduce data to only these sites 

```{r methmerge}
meth.quants.filtered.merged <- unite(meth.quants.filtered.norm, destrand = TRUE)

#batch 1

meth.quants.b1F1.filtered.merged <- unite(meth.quants.b1F1.filtered.norm, destrand = TRUE)
meth.quants.b1F2.filtered.merged <- unite(meth.quants.b1F2.filtered.norm, destrand = TRUE)
meth.quants.b1F3.filtered.merged <- unite(meth.quants.b1F3.filtered.norm, destrand = TRUE)


#batch 2
meth.quants.b2F1.filtered.merged <- unite(meth.quants.b2F1.filtered.norm, destrand = TRUE)
meth.quants.b2F2.filtered.merged <- unite(meth.quants.b2F2.filtered.norm, destrand = TRUE)
meth.quants.b2F3.filtered.merged <- unite(meth.quants.b2F3.filtered.norm, destrand = TRUE)
```

### Sample correlations

#### Correlation heatmap

- distance = pearson correlation
- agglomeration methods = ward

```{r correlations, fig.cap="Correlation matrix", fig.width = 8, fig.height = 8}
cor.mat  <- cor.meth(meth.quants.filtered.merged, method="pearson")
#gpairs(cor.mat, lower.pars=list(scatter="lm"), upper.pars=list(scatter="stats"), stat.pars=list(fontsize=16, use.color=FALSE), scatter.pars=list(pch=20, col="#00000022"), diag.pars=list(fontsize=10, show.hist=FALSE))

annotation_col <- data.frame(generation=metadata$generation, batch=factor(metadata$cohort))
rownames(annotation_col) <- metadata$samplename
pheatmap(cor.mat, clustering_method="ward.D2", annotation_col=annotation_col)

rm(cor.mat) ## cleanup
```

#### Dendrogram clustering

- distance = correlation
- methods = ward

```{r clustering, fig.cap="Methylation-based sample dendrogram", fig.width = 8, fig.height = 8}
meth.dendrogram <- clusterSamples(meth.quants.filtered.merged, dist="correlation", method="ward", plot=FALSE)
plot(meth.dendrogram)

rm(meth.dendrogram) ## cleanup
```

#### Principal Component Analysis

__All samples__

Overview of all included samples

```{r PCA.dimension.plot.all, fig.cap="Plot of First Two Principals Components", fig.width = 8, fig.height = 8}
PCASamples(meth.quants.filtered.merged)
```

Sample 14-b1-cntr is regarded as an outlier and removed from the analysis. 


```{dropoutlier}





__Per batch and generation__

Overview of each pair-wise comparison between control and treatment samples

```{r PCA.dimension.plot.batches, fig.cap="Plot of First Two Principals Components", fig.width = 16, fig.height = 24}

#batch 1
p1 <- PCASamples(meth.quants.b1F1.filtered.merged, main="Batch 1 - F1 & control")
p2 <- PCASamples(meth.quants.b1F2.filtered.merged, main="Batch 1 - F2 & control")
p3 <- PCASamples(meth.quants.b1F3.filtered.merged, main="Batch 1 - F3 & control")

#batch 2
p4 <- PCASamples(meth.quants.b2F1.filtered.merged, main="Batch 2 - F1 & control")
p5 <- PCASamples(meth.quants.b2F2.filtered.merged, main="Batch 2 - F2 & control")
p6 <- PCASamples(meth.quants.b2F3.filtered.merged, main="Batch 2 - F3 & control")

grid.arrange(p1,p4,p2,p5,p3,p6 , ncol=2)
```


## Differential methylation                           


### Logistic regression test

Significance wwas calculated by applying Logistic regression tests.
  
```{r logreg}
#my.diffMeth=calculateDiffMeth(meth.quants.filtered.norm.merged,slim=TRUE,weighted.mean=TRUE,num.cores=3)
#batch1

b1F1.diffMeth <- calculateDiffMeth(meth.quants.b1F1.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=2)
b1F2.diffMeth <- calculateDiffMeth(meth.quants.b1F2.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=2)
b1F3.diffMeth <- calculateDiffMeth(meth.quants.b1F3.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=2)

#batch2
b2F1.diffMeth <- calculateDiffMeth(meth.quants.b2F1.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=2)
b2F2.diffMeth <- calculateDiffMeth(meth.quants.b2F2.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=2)
b2F3.diffMeth <- calculateDiffMeth(meth.quants.b2F3.filtered.merged,slim=TRUE,  weighted.mean=TRUE,num.cores=2)

```


```{r fisher}
# pooled.methylBase <- pool(meth.quants.filtered.merged,sample.ids=c("control","F1"))
# my.diffMeth=calculateDiffMeth(pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)
# 
# #batch1
# b1F1.pooled.methylBase <- pool(meth.quants.b1F1.filtered.merged,sample.ids=c("control","F1"))
# b1F1.diffMeth <- calculateDiffMeth(b1F1.pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)
# 
# b1F2.pooled.methylBase <- pool(meth.quants.b1F2.filtered.merged,sample.ids=c("control","F2"))
# b1F2.diffMeth <- calculateDiffMeth(b1F2.pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)
# 
# b1F3.pooled.methylBase <- pool(meth.quants.b1F3.filtered.merged,sample.ids=c("control","F2"))
# b1F3.diffMeth <- calculateDiffMeth(b1F3.pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)
# 
# #batch2
# b2F1.pooled.methylBase <- pool(meth.quants.b2F1.filtered.merged,sample.ids=c("control","F1"))
# b2F1.diffMeth <- calculateDiffMeth(b2F1.pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)
# 
# b2F2.pooled.methylBase <- pool(meth.quants.b2F2.filtered.merged,sample.ids=c("control","F2"))
# b2F2.diffMeth <- calculateDiffMeth(b2F2.pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)
# 
# b2F3.pooled.methylBase <- pool(meth.quants.b2F3.filtered.merged,sample.ids=c("control","F2"))
# b2F3.diffMeth <- calculateDiffMeth(b2F3.pooled.methylBase,slim=TRUE,  weighted.mean=TRUE,num.cores=3)

```

### Differential methylation 

```{r methyldiffs}
# get differentially methylated regions with 25% difference and qvalue<0.01
diff.cutoff <- 25

#Diff25p <- get.methylDiff(my.diffMeth2,difference=diff.cutoff,qvalue=q.cutoff)
#Diff25pHypo <- get.methylDiff(my.diffMeth2,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
#Diff25pHyper <- get.methylDiff(my.diffMeth2,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

#batch1
b1F1.Diff25p <- get.methylDiff(b1F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b1F1.Diff25pHypo <- get.methylDiff(b1F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b1F1.Diff25pHyper <- get.methylDiff(b1F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b1F2.Diff25p <- get.methylDiff(b1F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b1F2.Diff25pHypo <- get.methylDiff(b1F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b1F2.Diff25pHyper <- get.methylDiff(b1F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b1F3.Diff25p <- get.methylDiff(b1F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b1F3.Diff25pHypo <- get.methylDiff(b1F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b1F3.Diff25pHyper <- get.methylDiff(b1F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

#batch2
b2F1.Diff25p <- get.methylDiff(b2F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b2F1.Diff25pHypo <- get.methylDiff(b2F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b2F1.Diff25pHyper <- get.methylDiff(b2F1.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b2F2.Diff25p <- get.methylDiff(b2F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b2F2.Diff25pHypo <- get.methylDiff(b2F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b2F2.Diff25pHyper <- get.methylDiff(b2F2.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")

b2F3.Diff25p <- get.methylDiff(b2F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff)
b2F3.Diff25pHypo <- get.methylDiff(b2F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hypo") 
b2F3.Diff25pHyper <- get.methylDiff(b2F3.diffMeth,difference=diff.cutoff,qvalue=q.cutoff,type="hyper")


```

```{r}
b1F1.gr=as(b1F1.Diff25p,"GRanges")
b1F2.gr=as(b1F2.Diff25p,"GRanges")
b1F3.gr=as(b1F3.Diff25p,"GRanges")

b2F1.gr=as(b2F1.Diff25p,"GRanges")
b2F2.gr=as(b2F2.Diff25p,"GRanges")
b2F3.gr=as(b2F3.Diff25p,"GRanges")

b1F1F2.overlap <- findOverlaps(b1F1.gr, b1F2.gr)
b1F1F3.overlap <- findOverlaps(b1F1.gr, b1F3.gr)

b1F1F2F3.overlap <- intersect(as.matrix(b1F1F2.overlap)[,1], as.matrix(b1F1F3.overlap)[,1])
b1F1F2F3.num <- length(b1F1F2F3.overlap)

b2F1F2.overlap <- findOverlaps(b2F1.gr, b2F2.gr)
b2F1F3.overlap <- findOverlaps(b2F1.gr, b2F3.gr)
b2F1F2F3.overlap <- intersect(as.matrix(b2F1F2.overlap)[,1], as.matrix(b2F1F3.overlap)[,1])
b2F1F2F3.num <- length(b2F1F2F3.overlap)

b1F1F2.num <- nrow(as.matrix(b1F1F2.overlap))
b1F1F3.num <- nrow(as.matrix(b1F1F3.overlap))
b2F1F2.num <- nrow(as.matrix(b2F1F2.overlap))
b2F1F3.num <- nrow(as.matrix(b2F1F3.overlap))

b1F1F2F3.coords <- getData(select(b1F1.Diff25p, b1F1F2F3.overlap))[,1:3]
b2F1F2F3.coords <- getData(select(b2F1.Diff25p, b2F1F2F3.overlap))[,1:3]

b1F1b2F1.overlap <- findOverlaps(b1F1.gr, b2F1.gr)
b1F1b2F1.num <- nrow(as.matrix(b1F1b2F1.overlap))

b1F1F2.idx <- as.matrix(b1F1F2.overlap)[,1]
b1F1F3.idx <- as.matrix(b1F1F3.overlap)[,1]
b1F1F2F3.idx <- b1F1F2F3.overlap

b2F1F2.idx <- as.matrix(b2F1F2.overlap)[,1]
b2F1F3.idx <- as.matrix(b2F1F3.overlap)[,1]
b2F1F2F3.idx <- b2F1F2F3.overlap



```

Comparing the overlap methylated positions between the different generations gives the following results..

__Batch 1__

* Overlap betwen F1 and F2: `r b1F1F2.num` nucleotides
* Overlap betwen F1 and F3: `r b1F1F3.num` nucleotides
* Overlap betwen F1, F2  and F3: `r b1F1F2F3.num` nucleotides

__Batch 2__

* Overlap betwen F1 and F2: `r b2F1F2.num` nucleotides
* Overlap betwen F1 and F3: `r b2F1F3.num` nucleotides
* Overlap betwen F1, F2  and F3: `r b2F1F2F3.num` nucleotides

__combined__

* Overlap betwen batch 1 F1 and batch 2 F1: `r b1F1b2F1.num` nucleotides





# Export

Loading genes and CpG islands. Promomoters are defined as 5000bp upstreams. Shores are 200bp flanks of CpG islands. 

```{r loadanno}
# annotation by genes
gene.obj=read.transcript.features(file.path(metaDir, "mm10_refseq.bed"), up.flank = 5000, down.flank = 0)

# annotation by CpG islands
cpg.obj=read.feature.flank(location=file.path(metaDir, "mm10_CpG_Islands.bed") ,feature.flank.name=c("CpGi","shores"), flank=2000)


```




## Result files

```{r results}
#batch1
write.table(annodiff(b1F1.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T), file= file.path(resultsDir, "b1F1.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b1F2.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T), file= file.path(resultsDir, "b1F2.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b1F2.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T), file= file.path(resultsDir, "b1F3.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)

write.table(annodiff(b1F1.Diff25p, b1F1F2.idx, gene.obj, cpg=cpg.obj), file= file.path(resultsDir, "b1F1F2.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b1F1.Diff25p, b1F1F3.idx, gene.obj, cpg=cpg.obj), file= file.path(resultsDir, "b1F1F3.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b1F1.Diff25p, b1F1F2F3.idx, gene.obj, cpg=cpg.obj), file= file.path(resultsDir, "b1F1F2F3.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)


#batch2
write.table(annodiff(b2F1.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T), file= file.path(resultsDir, "b2F1.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b2F2.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T), file= file.path(resultsDir, "b2F2.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b2F3.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T), file= file.path(resultsDir, "b2F3.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)

write.table(annodiff(b2F1.Diff25p, b2F1F2.idx, gene.obj, cpg=cpg.obj), file= file.path(resultsDir, "b2F1F2.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b2F1.Diff25p, b2F1F3.idx, gene.obj, cpg=cpg.obj), file= file.path(resultsDir, "b2F1F3.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(annodiff(b2F1.Diff25p, b2F1F2F3.idx, gene.obj, cpg=cpg.obj), file= file.path(resultsDir, "b2F1F2F3.Diff25p-results.xls"), quote=F, sep="\t", row.names=F, col.names=T)

```

__Batch 1__

* [Significant nucleotides b1F1](results/b1F1.Diff25p-results.xls)
* [Significant nucleotides b1F2](results/b1F2.Diff25p-results.xls)
* [Significant nucleotides b1F3](results/b1F3.Diff25p-results.xls)

* [Significant nucleotides common for b1F1 and b1F2](results/b1F1F2.Diff25p-results.xls)
* [Significant nucleotides common for b1F1 and b1F3](results/b1F1F3.Diff25p-results.xls)
* [Significant nucleotides common for b1F1, b1F2 and b1F3](results/b1F1F2F3.Diff25p-results.xls)

__Batch 2__

* [Significant nucleotides b2F1](results/b2F1.Diff25p-results.xls)
* [Significant nucleotides b2F2](results/b2F2.Diff25p-results.xls)
* [Significant nucleotides b2F3](results/b2F3.Diff25p-results.xls)

* [Significant nucleotides common for b2F1 and b2F2](results/b2F1F2.Diff25p-results.xls)
* [Significant nucleotides common for b2F1 and b2F3](results/b2F1F3.Diff25p-results.xls)
* [Significant nucleotides common for b2F1, b2F2 and b2F3](results/b2F1F2F3.Diff25p-results.xls)

## bed

bed-files with differentially methylated positions.

```{r bed}
write.table(getData(b1F1.Diff25p)[,1:3], file= file.path(resultsDir, "b1F1.Diff25p.bed"), quote=F, sep="\t", row.names=F, col.names=F)
write.table(getData(b1F2.Diff25p)[,1:3], file= file.path(resultsDir, "b1F2.Diff25p.bed"), quote=F, sep="\t", row.names=F, col.names=F)
write.table(getData(b1F3.Diff25p)[,1:3], file= file.path(resultsDir, "b1F3.Diff25p.bed"), quote=F, sep="\t", row.names=F, col.names=F)

write.table(getData(b2F1.Diff25p)[,1:3], file= file.path(resultsDir, "b2F1.Diff25p.bed"), quote=F, sep="\t", row.names=F, col.names=F)
write.table(getData(b2F2.Diff25p)[,1:3], file= file.path(resultsDir, "b2F2.Diff25p.bed"), quote=F, sep="\t", row.names=F, col.names=F)
write.table(getData(b2F3.Diff25p)[,1:3], file= file.path(resultsDir, "b2F3.Diff25p.bed"), quote=F, sep="\t", row.names=F, col.names=F)
```


* [b1F1.Diff25p.bed](results/b1F1.Diff25p.bed)
* [b1F2.Diff25p.bed](results/b1F2.Diff25p.bed)
* [b1F3.Diff25p.bed](results/b1F3.Diff25p.bed)
* [b2F1.Diff25p.bed](results/b2F1.Diff25p.bed)
* [b2F2.Diff25p.bed](results/b2F2.Diff25p.bed)
* [b2F3.Diff25p.bed](results/b2F3.Diff25p.bed)




## bedgraph

bed-files with differentially methylated positions and difference compared to control.

```{r bedgraph}
bedgraph(b1F1.Diff25p, file.path(resultsDir, "b1F1.Diff25p.bedgraph"), "meth.diff")
bedgraph(b1F2.Diff25p, file.path(resultsDir, "b1F2.Diff25p.bedgraph"), "meth.diff")
bedgraph(b1F3.Diff25p, file.path(resultsDir, "b1F3.Diff25p.bedgraph"), "meth.diff")

bedgraph(b2F1.Diff25p, file.path(resultsDir, "b2F1.Diff25p.bedgraph"), "meth.diff")
bedgraph(b2F2.Diff25p, file.path(resultsDir, "b2F2.Diff25p.bedgraph"), "meth.diff")
bedgraph(b2F3.Diff25p, file.path(resultsDir, "b2F3.Diff25p.bedgraph"), "meth.diff")

```

* [b1F1.Diff25p.bedgraph](results/b1F1.Diff25p.bedgraph)
* [b1F2.Diff25p.bedgraph](results/b1F2.Diff25p.bedgraph)
* [b1F3.Diff25p.bedgraph](results/b1F3.Diff25p.bedgraph)
* [b2F1.Diff25p.bedgraph](results/b2F1.Diff25p.bedgraph)
* [b2F2.Diff25p.bedgraph](results/b2F2.Diff25p.bedgraph)
* [b2F3.Diff25p.bedgraph](results/b2F3.Diff25p.bedgraph)



# Visualization  

## Differentially methylated Cs with promoter/exon/intron

__Batch 1__

```{r pei.plot.1, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.pei(b1F1.Diff25p,gene.obj, "b1F1")
p2 <- plot.targetanno.pei(b1F2.Diff25p,gene.obj, "b1F2")
p3 <- plot.targetanno.pei(b1F3.Diff25p,gene.obj, "b1F3")

grid.arrange(p1, p2, p3, ncol=2)
```

__Batch 2__

```{r pei.plot.2, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.pei(b2F1.Diff25p,gene.obj, "b2F1")
p2 <- plot.targetanno.pei(b2F2.Diff25p,gene.obj, "b2F2")
p3 <- plot.targetanno.pei(b2F3.Diff25p,gene.obj, "b2F3")

grid.arrange(p1, p2, p3, ncol=2)
```


## Differentially methylated Cs with CpG islands/shores

__Batch 1__

```{r cpgi.plot.1, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.cpgi(b1F1.Diff25p,cpg.obj, "b1F1")
p2 <- plot.targetanno.cpgi(b1F2.Diff25p,cpg.obj, "b1F2")
p3 <- plot.targetanno.cpgi(b1F3.Diff25p,cpg.obj, "b1F3")

grid.arrange(p1, p2, p3, ncol=2)
```


__Batch 2__

```{r cpgi.plot.2, fig.width = 16, fig.height = 16 }

p1 <- plot.targetanno.cpgi(b2F1.Diff25p,cpg.obj, "b2F1")
p2 <- plot.targetanno.cpgi(b2F2.Diff25p,cpg.obj, "b2F2")
p3 <- plot.targetanno.cpgi(b2F3.Diff25p,cpg.obj, "b2F3")

grid.arrange(p1, p2, p3, ncol=2)
```

### Number of hyper/hypo methylated regions/bases per chromosome

__Batch 1__

```{r diffmethperchr.1}
#diffMethPerChr(myDiff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL)

p1 <- diffMethPerChr(b1F1.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b1F1")
p2 <- diffMethPerChr(b1F2.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b1F2")
p3 <- diffMethPerChr(b1F3.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b1F3")

grid.arrange(p1, p2, p3, ncol=2)
```

__Batch 2__

```{r diffmethperchr.2}

p1 <- diffMethPerChr(b2F1.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b2F1")
p2 <- diffMethPerChr(b2F2.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b2F2")
p3 <- diffMethPerChr(b2F3.Diff25p, plot=T,qvalue.cutoff=q.cutoff, meth.cutoff=diff.cutoff,exclude=NULL, main="b2F3")

grid.arrange(p1, p2, p3, ncol=2)

```


### Manhattan plots

Corrected p-values (-log10(p)) plotted along the chromosomes. Values higher than 20 are showed in the upper part of the plot.

```{r manhattan.plot, fig.width = 16, fig.height = 24}
ymax=20

mp1 <- manhattan.plot (b1F1.Diff25p, "b1F1", ymax=ymax)
mp2 <- manhattan.plot (b1F2.Diff25p, "b1F2", ymax=ymax)
mp3 <- manhattan.plot (b1F3.Diff25p, "b1F3", ymax=ymax)

mp4 <- manhattan.plot (b2F1.Diff25p, "b2F1", ymax=ymax)
mp5 <- manhattan.plot (b2F2.Diff25p, "b2F2", ymax=ymax)
mp6 <- manhattan.plot (b2F3.Diff25p, "b2F3", ymax=ymax)

grid.arrange(mp1@ggplot, mp4@ggplot, mp2@ggplot, mp5@ggplot, mp3@ggplot, mp6@ggplot, ncol=2)
```


## Over-representated categories

Extracting genes for all methylated positions in promotors and checked for over-represented categories (GO, Reactome, KEGG, etc.)

```{r gprofiler}
#Batch 1
b1F1.Diff25p.prom <- unique(as.character(annodiff(b1F1.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b1F1.Diff25p.prom<- gprofiler(b1F1.Diff25p.prom, organism = "mmusculus")

b1F2.Diff25p.prom <- unique(as.character(annodiff(b1F2.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b1F2.Diff25p.prom<- gprofiler(b1F2.Diff25p.prom, organism = "mmusculus")

b1F3.Diff25p.prom <- unique(as.character(annodiff(b1F3.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b1F3.Diff25p.prom<- gprofiler(b1F3.Diff25p.prom, organism = "mmusculus")

b1F1F2.Diff25p.prom <- unique(as.character(annodiff(b1F1.Diff25p, b1F1F2.idx, gene.obj, cpg=cpg.obj) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b1F1F2.Diff25p.prom<- gprofiler(b1F1F2.Diff25p.prom, organism = "mmusculus")

b1F1F3.Diff25p.prom <- unique(as.character(annodiff(b1F1.Diff25p, b1F1F3.idx, gene.obj, cpg=cpg.obj) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b1F1F3.Diff25p.prom<- gprofiler(b1F1F3.Diff25p.prom, organism = "mmusculus")

b1F1F2F3.Diff25p.prom <- unique(as.character(annodiff(b1F1.Diff25p, b1F1F2F3.idx, gene.obj, cpg=cpg.obj) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b1F1F2F3.Diff25p.prom<- gprofiler(b1F1F2F3.Diff25p.prom, organism = "mmusculus")

#Batch 2
b2F1.Diff25p.prom <- unique(as.character(annodiff(b2F1.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b2F1.Diff25p.prom<- gprofiler(b2F1.Diff25p.prom, organism = "mmusculus")

b2F2.Diff25p.prom <- unique(as.character(annodiff(b2F2.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b2F2.Diff25p.prom<- gprofiler(b2F2.Diff25p.prom, organism = "mmusculus")

b2F3.Diff25p.prom <- unique(as.character(annodiff(b2F3.Diff25p, gene=gene.obj, cpg=cpg.obj, all=T) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b2F3.Diff25p.prom<- gprofiler(b2F3.Diff25p.prom, organism = "mmusculus")

b2F1F2.Diff25p.prom <- unique(as.character(annodiff(b2F1.Diff25p, b2F1F2.idx, gene.obj, cpg=cpg.obj) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b2F1F2.Diff25p.prom<- gprofiler(b2F1F2.Diff25p.prom, organism = "mmusculus")

b2F1F3.Diff25p.prom <- unique(as.character(annodiff(b2F1.Diff25p, b2F1F3.idx, gene.obj, cpg=cpg.obj) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b2F1F3.Diff25p.prom<- gprofiler(b2F1F3.Diff25p.prom, organism = "mmusculus")

b2F1F2F3.Diff25p.prom <- unique(as.character(annodiff(b2F1.Diff25p, b2F1F2F3.idx, gene.obj, cpg=cpg.obj) %>% filter(prom == 1) %>% collect %>% .[["symbol"]]))
go.b2F1F2F3.Diff25p.prom<- gprofiler(b2F1F2F3.Diff25p.prom, organism = "mmusculus")


```


```{r gowrite}
#batch1
write.table(go.b1F1.Diff25p.prom, file= file.path(resultsDir, "b1F1.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b1F2.Diff25p.prom, file= file.path(resultsDir, "b1F2.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b1F3.Diff25p.prom, file= file.path(resultsDir, "b1F3.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)

write.table(go.b1F1F2.Diff25p.prom, file= file.path(resultsDir, "b1F1F2.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b1F1F3.Diff25p.prom, file= file.path(resultsDir, "b1F1F3.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b1F1F2F3.Diff25p.prom, file= file.path(resultsDir, "b1F1F2F3.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)


#batch2
write.table(go.b2F1.Diff25p.prom, file= file.path(resultsDir, "b2F1.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b2F2.Diff25p.prom, file= file.path(resultsDir, "b2F2.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b2F3.Diff25p.prom, file= file.path(resultsDir, "b2F3.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)

write.table(go.b2F1F2.Diff25p.prom, file= file.path(resultsDir, "b2F1F2.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b2F1F3.Diff25p.prom, file= file.path(resultsDir, "b2F1F3.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(go.b2F1F2F3.Diff25p.prom, file= file.path(resultsDir, "b2F1F2F3.Diff25p-categories.xls"), quote=F, sep="\t", row.names=F, col.names=T)


```


__Batch 1__

* [Significant categories b1F1](results/b1F1.Diff25p-categories.xls)
* [Significant categories b1F2](results/b1F2.Diff25p-categories.xls)
* [Significant categories b1F3](results/b1F3.Diff25p-categories.xls)

* [Significant categories for positions common in b1F1 and b1F2](results/b1F1F2.Diff25p-categories.xls)
* [Significant categories for positions common in b1F1 and b1F3](results/b1F1F3.Diff25p-categories.xls)
* [Significant categories for positions common in b1F1, b1F2 and b1F3](results/b1F1F2F3.Diff25p-categories.xls)

__Batch 2__

* [Significant categories b2F1](results/b2F1.Diff25p-categories.xls)
* [Significant categories b2F2](results/b2F2.Diff25p-categories.xls)
* [Significant categories b2F3](results/b2F3.Diff25p-categories.xls)

* [Significant categories for positions common in b2F1 and b2F2](results/b2F1F2.Diff25p-categories.xls)
* [Significant categories for positions common in b2F1 and b2F3](results/b2F1F3.Diff25p-categories.xls)
* [Significant categories for positions common in b2F1, b2F2 and b2F3](results/b2F1F2F3.Diff25p-categories.xls)



# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

---

