---
bibliography: "references.bib"
csl: "bioinformatics.csl"
title: "Anandasabpathy - Vaccina vaccination route effect on gene expression in immune derived cells"
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_depth: 3
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library("knitcitations")
cleanbib()
options("citation_format" = "pandoc")

clientname="Niroshana Anandasabapathy"
clientemail="niroananda@gmail.com"
labPI="Niroshana Anandasabapathy"
lablocation="BWH"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  

All MoGene 2.0 ST arrays were processed using the 'oligo' BioConductor package `r citep("10.1093/bioinformatics/btq431")`, quality-controlled with arrayQualityMetrics `r citep("10.1093/bioinformatics/btn647")`, normalized with RMA `r citep("10.1093/biostatistics/4.2.249")`. Differentially expressed genes were identified using limma `r citep("http://link.springer.com/chapter/10.1007%2F0-387-29362-0_23")`.

---

# Setup

## Variables
Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
if(file.exists("/Users/johnhutchinson/Work/projects/na_vaccinia_scar")) {
  baseDir <- "/Users/johnhutchinson/Work/projects/na_vaccinia_scar"
} else {
  baseDir <- "~/projects/na_vaccinia_scar"
}
dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # colorblind friendly palette
covarsfilename="metadata.csv" # tab delimited file describing samples
lowintensity.percentile=0.1
mad.quantile.cutoff=0.1
pvalue.cutoff=0.05
highlight.color="green"
lfc.cutoff=0.58496251
pct_threshold <- 0.5

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

## Libraries

Bioconductor `r citep("10.1038/nmeth.3252")` and R `r citep("http://cran.r-project.org/")` libraries used to process the data.

```{r libraries}
library(arrayQualityMetrics) # array quality control reports
library(oligo) # array utilities

library(pd.mogene.2.0.st) # array layout annotation
library(mogene20sttranscriptcluster.db) # array probe to gene annotations

library(ggdendro)
library(CHBUtils)
library(limma)
library(ggplot2)
library(pheatmap) # pretty heatmaps

library(readr)
library(dplyr)
library(tidyr)
library(reshape2)
library(gridExtra)

library(biomaRt)

library(pvca)
library(sva)

# use older version of biomart to allow homology changes
humanmart <- useDataset("hsapiens_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL", host="jul2016.archive.ensembl.org"))
mousemart <- useDataset("mmusculus_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL", host="jul2016.archive.ensembl.org"))

mouseattributes = listAttributes(mousemart)
mousefilters = listFilters(mousemart)

humanattributes = listAttributes(humanmart)
humanfilters = listFilters(humanmart)
```

## Functions
- some custom functions

```{r functions}
PCAplot.sd.eset <- function(eset=NULL,  title=NULL){
  eset.core <- exprs(eset)
  myPca.core <- prcomp(t(eset.core))
  # SD of components
  sdevdf <- data.frame(cbind(as.numeric(myPca.core$sdev),c(1:length(myPca.core$sdev))))
  sdevdf$prop <-  sdevdf$X1/sum(sdevdf$X1)
  sdevdf$cum <- cumsum(sdevdf$prop)
  ggplot(sdevdf, aes(x=X2, y=prop)) + 
    geom_point(size=4, color="red") + 
    scale_x_continuous('Component') + 
    scale_y_continuous('Standard Deviation') +
    ggtitle(title) +
    geom_line(data=sdevdf, aes(x=X2, y=cum))
}
heatmap_fn = function(data, metadata ,...) {
  pheatmap(data, annotation=metadata, clustering_method = "ward.D2",
           clustering_distance_cols = "correlation", ...)
}
```

---

# Import Data and Metadata

- load in phenotypes and array names from metadata file [metadata file](../meta/metadata.csv)
  - this file contains the names and descriptions of CEL files contained in the data directory 
- used these file names to load in the CEL files

```{r dataload, results='hide'}
covars <- read.csv(file.path(metaDir, covarsfilename)) ## simple comma separated  file with CEL file in second column  and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
covars <- covars %>% tbl_df() %>% 
  unite(., category, location, celltype, treatment, remove=FALSE) %>% 
  mutate(replicate = 1:n()) %>% 
  unite(., sampleid, category, replicate, remove=FALSE) %>% 
  unite(., celltype_location, celltype, location, sep="-", remove=FALSE) %>% 
  as.data.frame()

# this only works with the FULL path to the files
affyraw <- read.celfiles(file.path(dataDir, as.character(unlist(covars$celfile))))
pData(affyraw) <- covars 
sampleNames(affyraw) <- pData(affyraw)$array_code
validObject(affyraw)
```

## Sample metadata

```{r covars, results='asis'}
# Sample information table
output <- pData(affyraw)
row.names(output) <- as.character(seq(1:nrow(output)))
DT::datatable(output, rownames = FALSE)
```

---

# Pre-Processing and Qualilty Control

## Raw Data QC

- using arrayQualityMetrics library `r citep("10.1093/bioinformatics/btn647")`

```{r rawQC, eval=FALSE}
affyraw.eset <- ExpressionSet(assayData=exprs(affyraw))
pData(affyraw.eset) <- pData(affyraw)
data.prepped <- prepdata(affyraw.eset, do.logtransform = TRUE, intgroup=c("location","celltype", "treatment" ))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "report_raw"), arrayTable = pData(affyraw.eset), reporttitle = "Raw QC Report")
```

**[Raw Data QC Report](../results/report_raw/index.html)**

Arrays #32, 33, 21 and 57 are outliers from the boxplots.
57, 62 and 5 also seem off by density plot.
Many of these also look bad by MA-plot.

Dropped `r kable(output[c(5,21,32,33,57,62),])`.

```{r dropoutliers}
outlierids <- pData(affyraw)$array_code[c(5,21,32,33,57,62)]
affyraw <- affyraw[,which(!pData(affyraw)$array_code %in% outlierids)]
```


## Background Correct and Normalize

- using RMA `r citep("10.1093/biostatistics/4.2.249")`

- summarize probesets on the gene ('core') level

```{r normalize, results='hide'}
affynorm.core <- oligo::rma(affyraw,  target="core", background=TRUE, normalize=TRUE)
```

## Normalized Data QC
- using arrayQualityMetrics library

```{r normQC, eval=FALSE}
data.prepped <- prepdata(affynorm.core, do.logtransform = FALSE, intgroup=c("location","celltype", "treatment" ))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "report_rma_core"), arrayTable = pData(affynorm.core), reporttitle = "Normalized QC Report")
```

**[Normalized Data QC Report](../results/report_rma.core/index.html)**
 
Everything looks good now. You can start to see some separation between cDC and migDC derived cells.

---

# Batch exploration

There are a number of potential sources of batch effects (array batch, FACS date, etc). Here we explore whether they have an effect on the data and if we need to correct for them.

What are the major sources of variation in each sample set?

Which samples correlate the most strongly with each other? Do they cluster by batch?


## All celltypes {.tabset}

```{r batchsetup_all}
affynorm.sub <- affynorm.core
pd <- pData(affynorm.core) %>% tbl_df %>% dplyr::select(treatment, batch, arraydate,facsdate, RNA_prep_date, celltype, location, celltype_location) %>% as.data.frame()
```

### Sources of variability

- for all covariates that are present in more than one sample

```{r, varsource_all}
batch.factors <- c()

if(all(table(pd$celltype)>1)  & length(unique(pd$celltype))>1){
  batch.factors <- c(batch.factors, "celltype")
}
if(all(table(pd$location)>1)  & length(unique(pd$location))>1){
  batch.factors <- c(batch.factors, "location")
}
if(all(table(pd$celltype_location)>1) & length(unique(pd$celltype_location))>1){
  batch.factors <- c(batch.factors, "celltype_location")
}
if(all(table(pd$treatment)>1)& length(unique(pd$treatment))>1){
  batch.factors <- c(batch.factors, "treatment")
}
if(all(table(pd$batch)>1)& length(unique(pd$batch))>1){
  batch.factors <- c(batch.factors, "batch")
}
if(all(table(pd$arraydate)>1)& length(unique(pd$arraydate))>1){
  batch.factors <- c(batch.factors, "arraydate")
}

if(all(table(pd$facsdate)>1)& length(unique(pd$facsdate))>1){
  batch.factors <- c(batch.factors, "facsdate")
}
if(all(table(pd$RNA_prep_date)>1)& length(unique(pd$RNA_prep_date))>1){
  batch.factors <- c(batch.factors, "RNA_prep_date")
}
# allow it to work with just one batching factor
if (length(batch.factors)==1){batch.factors=c(batch.factors, batch.factors)}

if (!is.null(batch.factors)){
  pvcaObj <- pvcaBatchAssess(affynorm.sub, batch.factors, pct_threshold)
  plotme <- cbind(as.character(unlist(pvcaObj$label)), as.numeric(as.character(unlist(pvcaObj$dat)))) %>% tbl_df() %>% 
    magrittr::set_colnames(c("label", "value")) %>% mutate(value=as.numeric(value))
  ggplot(plotme, aes(x=label, y=value, fill=label))+geom_bar(stat="identity") +theme_bw()+theme(axis.text.x = element_text(angle = 65, hjust = 1))+ theme(legend.position="none")+ggtitle("Sources of variability")
}
```

The biggest source of variability in the data is the location source of the samples. Batch, celltype and treatment are also sources of variation. FACS date and RNA_prep_date also contribute to the variance in the data.

### Pearsons correlation

```{r, corheatmap_pearsons_all, fig.height=12}
heatmap_fn(data=cor(exprs(affynorm.sub), method="pearson"), metadata=pd)
```

### Spearmans correlation

```{r,corheatmap_spearman_all, fig.height=12}
heatmap_fn(data=cor(exprs(affynorm.sub), method="spearman"), metadata=pd)
```

### PCA 

```{r , pca_prep_all}
eset <- affynorm.sub
pd <- pData(eset)
pd$id <- row.names(pd)
eset.core <- exprs(eset)
PCA.core <- prcomp(t(eset.core))
PCA.core.sub <- as.data.frame(PCA.core$x[, 1:4])
PCA.core.sub$id <- row.names(PCA.core.sub)
plotme <- base::merge(PCA.core.sub, pd, by="id")
```

#### PC1 vs. PC2

```{r pca_1_2_all,  fig.height=12, fig.width=10}
p1 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=arraydate))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("Array date")
p2 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=batch))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ ggtitle("Array batch")
p3 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=facsdate))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("FACS date")
p4 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=RNA_prep_date))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("RNA prep date")
p5 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=location))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("location")
p6 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=celltype))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("celltype")
p7 <-  ggplot(plotme, aes(x=PC1, y=PC2, shape=celltype, color=treatment))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("treatment")

grid.arrange(p1,p2,p3,p4,p5,p6,p7,ncol=2)
```


Looking at all of the samples at once, it's clear there are some factors that cluster that samples best, but the picture is not entirely clear. To try to simplify matters, I ran the analyses for each cell type and location combination separately.

## CD8_plus-cDC {.tabset}

`r celltype="CD8_plus-cDC"`

```{r batchsetup_CD8_plus-cDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, arraydate,facsdate, RNA_prep_date, celltype, location, celltype_location) %>% as.data.frame()
```

### Sources of variability

```{r, varsource_CD8_plus-cDC, ref.label="varsource_all"}
```

### Pearsons correlation

```{r, corheatmap_pearsons_CD8_plus-cDC, fig.height=8, ref.label="corheatmap_pearsons_all"}
```

### Spearmans correlation

```{r,corheatmap_spearman_CD8_plus-cDC, ref.label="corheatmap_spearman_all", fig.height=8}
```

### PCA 

```{r , pca_prep_CD8_plus-cDC, ref.label="pca_prep_all"}
```

#### PC1 vs. PC2

```{r pca_1_2_CD8_plus-cDC, ref.label="pca_1_2_all", fig.height=12, fig.width=10}
```



## CD103_plus-migDC {.tabset}

`r celltype="CD103_plus-migDC"`

```{r batchsetup_CD103_plus-migDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, arraydate,facsdate, RNA_prep_date, celltype, location, celltype_location) %>% as.data.frame()
```

### Sources of variability

```{r, varsource_CD103_plus-migDC, ref.label="varsource_all"}
```

### Pearsons correlation

```{r, corheatmap_pearsons_CD103_plus-migDC, fig.height=8, ref.label="corheatmap_pearsons_all"}
```

### Spearmans correlation

```{r,corheatmap_spearman_CD103_plus-migDC, ref.label="corheatmap_spearman_all", fig.height=8}
```

### PCA

```{r , pca_prep_CD103_plus-migDC, ref.label="pca_prep_all"}
```

#### PC1 vs. PC2

```{r pca_1_2_CD103_plus-migDC, ref.label="pca_1_2_all", fig.height=12, fig.width=10}
```

## CD11B_minus-migDC {.tabset}

`r celltype="CD11B_minus-migDC"`

```{r batchsetup_CD11B_minus-migDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, arraydate,facsdate, RNA_prep_date, celltype, location, celltype_location) %>% as.data.frame()
```

### Sources of variability

```{r, varsource_CD11B_minus-migDC, ref.label="varsource_all"}
```

### Pearsons correlation

```{r, corheatmap_pearsons_CD11B_minus-migDC, fig.height=8, ref.label="corheatmap_pearsons_all"}
```

### Spearmans correlation

```{r,corheatmap_spearman_CD11B_minus-migDC, ref.label="corheatmap_spearman_all", fig.height=8}
```

### PCA 

```{r , pca_prep_CD11B_minus-migDC, ref.label="pca_prep_all"}
```

#### PC1 vs. PC2

```{r pca_1_2_CD11B_minus-migDC, ref.label="pca_1_2_all", fig.height=12, fig.width=10}
```

## CD11B_plus-migDC {.tabset}

`r celltype="CD11B_plus-migDC"`

```{r batchsetup_CD11B_plus-migDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, arraydate,facsdate, RNA_prep_date, celltype, location, celltype_location) %>% as.data.frame()
```

### Sources of variability

```{r, varsource_CD11B_plus-migDC, ref.label="varsource_all"}
```

### Pearsons correlation

```{r, corheatmap_pearsons_CD11B_plus-migDC, fig.height=8, ref.label="corheatmap_pearsons_all"}
```

### Spearmans correlation

```{r,corheatmap_spearman_CD11B_plus-migDC, ref.label="corheatmap_spearman_all", fig.height=8}
```

### PCA

```{r , pca_prep_CD11B_plus-migDC, ref.label="pca_prep_all"}
```

#### PC1 vs. PC2

```{r pca_1_2_CD11B_plus-migDC, ref.label="pca_1_2_all", fig.height=12, fig.width=10}
```


## CD11B_plus-cDC {.tabset}

`r celltype="CD11B_plus-cDC"`

```{r batchsetup_CD11B_plus-cDC}
affynorm.sub <- affynorm.core[,which(pData(affynorm.core)$celltype_location==celltype)]
pd <- pData(affynorm.sub) %>% tbl_df %>% dplyr::select(treatment, batch, arraydate,facsdate, RNA_prep_date, celltype, location, celltype_location) %>% as.data.frame()
```

### Sources of variability

```{r, varsource_CD11B_plus-cDC, ref.label="varsource_all"}
```

### Pearsons correlation

```{r, corheatmap_pearsons_CD11B_plus-cDC, fig.height=8, ref.label="corheatmap_pearsons_all"}
```

### Spearmans correlation

```{r,corheatmap_spearman_CD11B_plus-cDC, ref.label="corheatmap_spearman_all", fig.height=8}
```

### PCA

```{r , pca_prep_CD11B_plus-cDC, ref.label="pca_prep_all"}
```

#### PC1 vs. PC2

```{r pca_1_2_CD11B_plus-cDC, ref.label="pca_1_2_all", fig.height=12, fig.width=10}
```

Overall, these reasults indicate that the largest effect is from cell location, with batch/arraydate as the second largest. The other batch variables like the FACS date and RNA prep date are also likely to affect the differntial gene expression analysis. Unfortunately, because they are are not fully balanced with respect to the sample classes of interest, we can't adjust for them. We can however, adjust for the array batch, using the full dataset. As the RNA prep and FACS dates are somewhat confounded with the array batches, adjusting for array batch may help with them as well. 

# Batch correction

## ComBat

`r citep("doi.org/10.1093/biostatistics/kxj037")`

```{r combat}
# Create model with conditions of interest as factor variable
modcombat <- model.matrix(~treatment+celltype+location, data=pData(affynorm.core))
batch <- pData(affynorm.core)$batch
# Modify expression matrix
eset.core <- exprs(affynorm.core)
combat_edata <- ComBat(dat=eset.core,
                       batch=batch,
                       mod=modcombat,
                       par.prior=TRUE, 
                       prior.plots=TRUE)
affynorm.batch <- affynorm.core
exprs(affynorm.batch) <- combat_edata
```

## PCA Assessment {.tabset}

```{r PCAassess_fn}
PCA_grid <- function(affyeset=NULL){
  eset.pca <- affyeset
  pd.pca <- pData(eset.pca)
  pd.pca$id <- row.names(pd.pca)
  eset.pca <- exprs(eset.pca)
  PCA <- prcomp(t(eset.pca))
  PCA.sub <- as.data.frame(PCA$x[, 1:4])
  PCA.sub$id <- row.names(PCA.sub)
  plotme <- base::merge(PCA.sub, pd.pca, by="id")
  
  p1 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=arraydate))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("Array date")
  p2 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=batch))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ ggtitle("Array batch")
  p3 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=facsdate))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("FACS date")
  p4 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=RNA_prep_date))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("RNA prep date")
  p5 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=location))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("location")
  p6 <- ggplot(plotme, aes(x=PC1, y=PC2, shape=treatment, color=celltype))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("celltype")
  p7 <-  ggplot(plotme, aes(x=PC1, y=PC2, shape=celltype, color=treatment))+geom_point(size=3)+scale_shape_manual(values = c(15,16,17,18))+ggtitle("celltype")
  
  grid.arrange(p1,p2,p3,p4,p5,p6,p7,ncol=2)
}  
```

### All Cells

```{r PCApostbatch_all,fig.height=12, fig.width=10}
PCA_grid(affynorm.batch)
```

### 

### CD8_plus-cDC

```{r PCApostbatch_CD8_plus-cDC,fig.height=12, fig.width=10}
celltype <- "CD8_plus-cDC"
affynorm.batch.sub <- affynorm.batch[,which(pData(affynorm.batch)$celltype_location==celltype)]
PCA_grid(affynorm.batch.sub)
```


### CD103_plus-migDC

```{r PCApostbatch_CD103_plus-migDC,fig.height=12, fig.width=10}
celltype <- "CD103_plus-migDC"
affynorm.batch.sub <- affynorm.batch[,which(pData(affynorm.batch)$celltype_location==celltype)]
PCA_grid(affynorm.batch.sub)
```

### CD11B_minus-migDC

```{r PCApostbatch_CD11B_minus-migDC,fig.height=12, fig.width=10}
celltype <- "CD11B_minus-migDC"
affynorm.batch.sub <- affynorm.batch[,which(pData(affynorm.batch)$celltype_location==celltype)]
PCA_grid(affynorm.batch.sub)
```

### CD11B_plus-migDC

```{r PCApostbatch_CD11B_plus-migDC,fig.height=12, fig.width=10}
celltype <- "CD11B_plus-migDC"
affynorm.batch.sub <- affynorm.batch[,which(pData(affynorm.batch)$celltype_location==celltype)]
PCA_grid(affynorm.batch.sub)
```

### CD11B_plus-cDC

```{r PCApostbatch_CD11B_plus-cDC,fig.height=12, fig.width=10}
celltype <- "CD11B_plus-cDC"
affynorm.batch.sub <- affynorm.batch[,which(pData(affynorm.batch)$celltype_location==celltype)]
PCA_grid(affynorm.batch.sub)
```

There is less clustering by array batch, but we still aren't seeing cells cluster strongly by treatment.



---
# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```


---

# References

```{r writebib, results='hide', echo=FALSE, message=FALSE, cache=FALSE}
write.bibtex(file="references.bib")
```