---
title: "Perrimon Lab *D. melanogaster* blood scRNA-Seq"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

[bcbio umis]: http://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html#umis
[bcbio scrnaseq]: http://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html#single-cell-rna-sequencing
[`pear`]: http://www.exelixis-lab.org/web/software/pear
[`umis`]: https://github.com/vals/umis

# Overview

- Demultiplex Illumina NextSeq BCL files using `bcl2fastq`.
- Identify barcodes and unique molecular identifiers (UMIs). 
- Correct and filter barcodes.
- Align with kallisto and process transcript compatibility counts (TCCs).
- Perform quality control and clustering.

## inDrops pipeline history

v1 : original design where R2 is the biological read and R1 is the metadata read.
v2 : inversion of v1 where R1 is the biological read and R2 is the metadata read.
v3 : summer 2016 redesign requiring manual demultiplexing. R1 is the biological read. R2 carries the first half of the gel barcode. R3 carries the library index. R4 carries the second half of the gel barcode, the UMI, and a fraction of the polyA tail.

We are using v3.



# Questions

- What is the scientific question?
- Samples >90% viable for 30-60 min on ice?
- Analysis method?
- **Batch effects**: Biological, inDrops run, library preparation, days
- Is there sample remaining for qPCR? Precise quantitation is key!



# Sequencing V3 inDrops Library

These are the split FASTQ files with suffixes `_R1` through `_R4`, generated by `bcl2fastq`.

1. 61 bp Read 1: transcript
2. 8 bp Index Read 1 (i7): single cell barcode
3. 8 bp Index Read 2 (i5): library index
4. 14 bp Read 2: barcode/unique molecular identifiers (UMIs)



# Workflow

Demultiplex the raw BCL files with [`bcl2fastq`].
**Add the settings**

Combine the FASTQ files across lanes if necessary with `cat`.

Subset the IDX barcodes:

- idx-16 (M, other): TTATGCGA / TCGCATAA
- idx-17 (GFP): GGAGGTAA / TTACCTCC
- idx-18 (unsorted): CATAACTG / CAGTTATG

Let's run bcbio keeping only `idx-17` and `idx-18`

GFP+ (idx-17; GGAGGTAA)
“Unsorted” (idx-18; CATAACTG)

In sum, we got 2000 cells sequenced. 

"GFP+" means pure FAC Sorted immune cells while "Unsorted" means all cells (not sorted).

Note that unsorted will also have GFP. 

The third one is  from another group.
M  (TTATGCGA; idx-16)


Set up the `bcbio-nextgen` `scRNA-seq` template.




## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
