---
title: "Differential Expression Analysis"
output:
  html_document:
    toc: true
    toc_depth: 3
    theme: journal
    css: custom.css
author: "John Hutchinson"
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

RNAseq DE analysis for Shachar Dagan  (shachar_dagan@hms.harvard.edu),  Gu group at HMS. Using new set of LE samples to replace LE2, LE3 and LE4. 

Contact John Hutchinson (jhutchin@hsph.harvard.edu) for additional details.

The most recent update of this html document occurred: `r date()`.

The sections below provide code to reproduce the included results and plots. 

---

# Setup

## Libraries and Variables

```{r vars}
library(ggplot2)
#library(gplots)
library(DT)
library(CHBUtils)
library(DESeq2)
#library(Biobase)
library(gProfileR)
library(pheatmap)
library(dplyr)
library(biomaRt)
library(gProfileR)
library(RUVSeq)

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442","#0072B2", "#D55E00", "#CC79A7")

baseDir <- "~/Work/projects/gu_rnaseq/"
dataDir <- file.path(baseDir, "results/final/2016-06-15_project")
resultsDir <- file.path(baseDir, "results/final/2016-06-15_project/report")
```

---

# Data and Metadata Import 

```{r import}
project_summary = file.path(dataDir, "project-summary.csv")
counts_file = file.path(dataDir, "combined.counts")

summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE) 
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
summarydata$fastqcode <- NULL
summarydata$phenotype <- NULL

counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality","rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate","Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped","rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.","Genes.Detected", "Unique.Starts.Per.Read","unique_starts_per_read","complexity", "X5.3.bias", "Duplicates.pct", "Duplicates", "Mapped.reads","Median.insert.size", "Mapped.reads.pct","Total.reads","avg_coverage_per_region", "Mapped.Reads", "Average.insert.size")

metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
metadata$day <- as.character(metadata$day)
metadata$batch <- as.character(metadata$batch)
summarydata$day <- as.character(summarydata$day)
```

---

# Custom Functions

```{r heatmapfunction}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```


---

# Data Manipulations

## Remove rRNA and mitochondrial genes

```{r rRNA removal}
ensemblmart = useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl", host = "jul2015.archive.ensembl.org")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)

conversions = getBM(attributes = c("ensembl_gene_id", "mgi_symbol", "gene_biotype", "chromosome_name"), mart =ensemblmart)

# id rRNA genes by biotype coding
rrna_biotypes = c("rRNA", "Mt_rRNA", "misc_RNA", "snRNA", "snoRNA", "tRNA", "Mt_tRNA")
rrna_genes <- unique(subset(conversions, gene_biotype %in% rrna_biotypes)$ensembl_gene_id)

#id mitochondrial proteing coding genes by chromosome
mt_genes <- unique(subset(conversions, conversions$chromosome_name=="MT" & conversions$gene_biotype=="protein_coding")$ensembl_gene_id)

# Analysis without mt and rRNA genes
counts <- counts[!rownames(counts) %in% rrna_genes & !rownames(counts) %in% mt_genes, ]
```

## Drop Outlier Samples

Previous QC analysis showed that LE2, LE3 and LE4 are all outliers with low mapping rates. Dropped these samples from the analysis.

```{r dropoutliers}
metadata <- metadata[!row.names(metadata) %in% c("LE2_1", "LE3_1", "LE4_1"),]
counts <- counts[,!names(counts) %in% c("LE2_1", "LE3_1", "LE4_1")]
```

## Remove genes with no counts for any of the remaining samples

```{r dropzeros}
counts <- counts[!apply(counts, 1, function(x) all(x==0)),]
```

## Import raw counts into DESeq2

```{r DEobject}
eset <- new("ExpressionSet", exprs = as.matrix(counts))
metadata <- metadata[order(metadata$sampletype, metadata$day),]
pData(eset) <- metadata
validObject(eset)

dds <- DESeqDataSetFromMatrix(countData = exprs(eset), colData = pData(eset), design = ~day+batch+sampletype)
dds <- DESeq(dds)
```

## PCA plot of normalized, variant stabilized data

```{r PCAs}
rld = rlog(dds)
plotPCA(rld, intgroup=c("batch", "sampletype"))
plotPCA(rld, intgroup=c("batch", "sampletype", "day"))
```

There is a batch effect, which we will need to eliminate as it is confounded with the sample class. Ideally, all the lung samples would cluster together. 


Try using RUVseq to remove variation associated wtih batch.

First, try leveraging the two technical replicates we have that span the batches.

### RUVseq using replicates

```{r ruvseqs}
#RUVs using replicates
## specify replicates
replicates <- matrix(data=c(c(1,2), c(6,7)), byrow=TRUE, nrow=2)
## run RUVseq on size-adjusted data from DESeq2
batch_ruvs_calc <- RUVs(as.matrix(counts, normalize=TRUE), rownames(counts), k=1, replicates, isLog=FALSE)
## variance stabilize RUVseq adjusted results
normed.ruvs.vs <- rlog(batch_ruvs_calc$normalizedCounts)
## PCA plot the variaznce stabilized RUVseqS adjusted results
se.ruvs <- SummarizedExperiment(normed.ruvs.vs, colData=colData(dds))
plotPCA(DESeqTransform(se.ruvs), intgroup=c("sampletype", "batch"))
plotPCA(DESeqTransform(se.ruvs), intgroup=c("sampletype", "batch", "day"))

# this is better but doesn't completely remove the batch effect (note the two lung samples off by themselves are the ones that are replicated)
```

This is an improvement over the uncorrected data but it still doesn't completely remove the batch effect (note the two lung samples off by themselves are the ones that are replicated)

### RUVseq using residuals

```{r RUVr}
## RUVr using residuals
# setup a design that ignores batch
design <- model.matrix(~day+sampletype, data=as.data.frame(colData(dds)))
# use the EdgeR GLM to find residuals using this design
y <- DGEList(counts=counts(dds), group=colData(dds)$sampletype)
y <- calcNormFactors(y, method="TMM")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)

fit <- glmFit(y, design)
res <- residuals(fit, type="deviance")

# plot PCAs of residuals to see if they capture day and batch
se <- SummarizedExperiment(res, colData=colData(dds))
plotPCA(DESeqTransform(se), intgroup=c( "day"))
plotPCA(DESeqTransform(se), intgroup=c( "batch"))
# some amount of batch captured

batch_ruvr_calc <- RUVr(counts(dds), rownames(counts(dds)), k=1, res) 
normed.ruvr.vs <- rlog(batch_ruvr_calc$normalizedCounts)
## PCA plot the variaznce stabilized RUVseqR adjusted results
se.ruvr <- SummarizedExperiment(normed.ruvr.vs, colData=colData(dds))
plotPCA(DESeqTransform(se.ruvr), intgroup=c("sampletype", "batch"))
plotPCA(DESeqTransform(se.ruvr), intgroup=c("sampletype", "batch", "day"))

```

Now appears that batch and sampletype are confounded...

---

Try out some methods and look at heatmaps to see how well they work.

a) Try DESeq2 with replicate corrected data

b) Will also try just dropping the 2 new replicated samples and include batch as a covariate in DESeq2

```{r, DE1}
library(DESeq2)

dds.ruvs <- DESeqDataSetFromMatrix(countData = exprs(eset),
                                   colData = cbind(batch_ruvs_calc$W, pData(eset)),
                                   design = ~W_1+sampletype)
dds <- DESeq(dds) 

results.dr <- results(dds, alpha=0.01, lfcThreshold = 1)
results.df <- as.data.frame(results.dr)
results.df.annot <- annotate_df(row2colnames(results.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")

DEresults.df.annot <- subset(results.df.annot,padj<0.01 & abs(log2FoldChange)>1 )
# reoder stats by logFC and pull out top  results
top.results.df.annot <- DEresults.df.annot[order(abs(DEresults.df.annot$log2FoldChange)),][1:100,]
# grab the ensemblids of the top reuslts
top.ensemblids <- as.character(top.results.df.annot$ensemblid)
# pull variance stabilizied data into a dataframe
rld.df <- assay(rld[,which(!colnames((eset)) %in% c("LE1_2", "LE5_2"))])
# subset data to top genes
top.rld.df <- rld.df[row.names(rld.df) %in% top.ensemblids,]
# get gene symbols
top.rld.df.annot <-  annotate_df(row2colnames(top.rld.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")

top.rld.df.annot <- top.rld.df.annot[match(row.names(top.rld.df), top.rld.df.annot$ensemblid),]
identical(as.character(top.rld.df.annot$ensemblid), row.names(top.rld.df))

heatmap_fn(top.rld.df, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8, scale="row")

heatmap_fn(top.rld.df, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8)

```

Run DESeq2 with batch as covariate in design.  
Keep pairing information.   
Drop new, replicated samples.  

```{r, DE2}
# drop repeat samples
eset.sub <- eset[,which(!sampleNames((eset)) %in% c("LE1_2", "LE5_2"))]

dds <- DESeqDataSetFromMatrix(countData = exprs(eset.sub), colData = pData(eset.sub), design = ~batch+day+sampletype)
dds <- DESeq(dds)


results.dr <- results(dds, alpha=0.01, lfcThreshold = 1)
results.df <- as.data.frame(results.dr)
results.df.annot <- annotate_df(row2colnames(results.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")

DEresults.df.annot <- subset(results.df.annot,padj<0.01 & abs(log2FoldChange)>1 )
# reoder stats by logFC and pull out top  results
top.results.df.annot <- DEresults.df.annot[order(abs(DEresults.df.annot$log2FoldChange)),][1:100,]
# grab the ensemblids of the top reuslts
top.ensemblids <- as.character(top.results.df.annot$ensemblid)
# pull variance stabilizied data into a dataframe
rld.df <- assay(rld[,which(!colnames((eset)) %in% c("LE1_2", "LE5_2"))])
# subset data to top genes
top.rld.df <- rld.df[row.names(rld.df) %in% top.ensemblids,]
# get gene symbols
top.rld.df.annot <-  annotate_df(row2colnames(top.rld.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")

top.rld.df.annot <- top.rld.df.annot[match(row.names(top.rld.df), top.rld.df.annot$ensemblid),]
identical(as.character(top.rld.df.annot$ensemblid), row.names(top.rld.df))

heatmap_fn(top.rld.df, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8, scale="row")

heatmap_fn(top.rld.df, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8)
```

Run DESeq2 with batch as covariate in design.   
Discard pairing information.   
Drop new, replicated samples.  

```{r, DE3}
# drop repeat samples
eset.sub <- eset[,which(!sampleNames((eset)) %in% c("LE1_2", "LE5_2"))]

dds <- DESeqDataSetFromMatrix(countData = exprs(eset.sub), colData = pData(eset.sub), design = ~batch+sampletype)
dds <- DESeq(dds)

results.dr <- results(dds, alpha=0.01, lfcThreshold = 1)
results.df <- as.data.frame(results.dr)
results.df.annot <- annotate_df(row2colnames(results.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")

DEresults.df.annot <- subset(results.df.annot,padj<0.01 & abs(log2FoldChange)>1 )
# reoder stats by logFC and pull out top  results
top.results.df.annot <- DEresults.df.annot[order(abs(DEresults.df.annot$log2FoldChange)),][1:100,]
# grab the ensemblids of the top reuslts
top.ensemblids <- as.character(top.results.df.annot$ensemblid)
# pull variance stabilizied data into a dataframe
rld.df <- assay(rld[,which(!colnames((eset)) %in% c("LE1_2", "LE5_2"))])
# subset data to top genes
top.rld.df <- rld.df[row.names(rld.df) %in% top.ensemblids,]
# get gene symbols
top.rld.df.annot <-  annotate_df(row2colnames(top.rld.df, "ensemblid"),'ensemblid', 'mmusculus_gene_ensembl', "ensembl_gene_id", "mgi_symbol")

top.rld.df.annot <- top.rld.df.annot[match(row.names(top.rld.df), top.rld.df.annot$ensemblid),]
identical(as.character(top.rld.df.annot$ensemblid), row.names(top.rld.df))

heatmap_fn(top.rld.df, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8, scale="row")

heatmap_fn(top.rld.df, labels_row=top.rld.df.annot$mgi_symbol, fontsize_row=8)
```

