```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120))
options(width=200)
```

# STANDARD MICROARRAY ANALYSES OF U133PLUS2.0 STUDIES

---

Analysis of public datasets for differential gene expression signatures between adults and children for [Les Kobzik](mailto:LKOBZIK@hsph.harvard.edu) as part of the DARPA 7-day Biodefense Program.

Contact [John Hutchinson](mailto:jhutchin@hsph.harvard.edu) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

----

# GENERAL SETUP

## General purpose R libraries necessary for analysis

```{r general_libraries}
source("http://bioconductor.org/biocLite.R") # BioConductor script necessary for installing new BioC libraries with biocLite()
library(plyr) # library for iteratively working with data structures
library(ggplot2) # library for plotting 
library(RColorBrewer) # library for generating color palettes
library(googleVis) # library for presenting tables
```

## Locations of directories for data, metadata and results

```{r general_directories}
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/lk_FOY")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/lk_FOY"
  }  else if (file.exists("/Volumes/home08/jhutchin/consults/lk_FOY/")) {
    baseDir <- "/Volumes/home08/jhutchin/consults/lk_FOY"
    } else {
      baseDir <- "/Volumes/ody/consults/lk_FOY"
      }
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
````

----

# Microarray Analysis

## Load [Bioconductor][1] libraries for the microarray analysis 

```{r microarray_analysis_libraries}
library(virtualArray)
library(beadchip)
library(lumi)
# to parse the CEL files and work with intensity values
library(affy) 
# for QC reports
library(arrayQualityMetrics)
# library to do stats 
library(limma) 
# pretty heatmaps
library(pheatmap) 
# annotations for the hgU1332.0Plus array
library(hgu133plus2.db) 
library(hgu133a2.db)
library(hgu133a.db)
library(illuminaHumanv3.db)
library(illuminaHumanv4.db)
```

- use virtualArray library to  bring in as many array types as possible
- used arrays for which there are bioconductor array platform annotations
- trying out import of the Illumina arrays for now, as I need to work out how to import and work with them

Platforms then include HT12v3 and HT12v4 Arrays (essentially, this is an Illumina dataset)

- also subsetted to samples for which we gender, age and raw data

(also excluded the studies I excluded from the previous U133Plus2.0 analyses)

```{r subset_data}
refined.metadata <- read.delim(file.path(metaDir,"unified-metadata-refined.04-17-13.tab"))
refined.metadata <- refined.metadata[which(!is.na(refined.metadata$age) & !is.na(refined.metadata$gender) & !is.na(refined.metadata$CEL_regex)),]
refined.metadata  <- refined.metadata[!(refined.metadata$study %in% c("GSE18897", "MEXP-2917", "TABM-940", "TABM-666")),]
refined.metadata  <- refined.metadata[refined.metadata$platform %in% c("GPL6947", "GPL10558"),]
HT.meta <- refined.metadata
write.table(HT.meta, file.path(metaDir, "unified-metadata-refined_HT.tab"), quote=F, sep="\t", row.names=F, col.names=T)
write.table(HT.meta, file.path(metaDir, "unified-metadata-refined_HT.xls"), sep="\t", row.names=F, col.names=T)
```

That metadata file is [here](./meta/unified-metadata-refined_HT.xls)

## Setup

### Setup some more variables for the microarray analyses
- color palette for plots
- age ranges and names of the groups  
  - neonate = 0-4 years
  - child = 5-10 years
  - adult = 18-40 years

```{r microarray_analysis_variables}
# colorblind friendly palette
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
# age ranges to compare
child.age.range <- c(5,10)
adult.age.range <- c(18,40)
neonate.age.range <- c(0,4)
```

### Load and parse the unified, refined and subsetted metadata file 
To make a covariate table that allows you to load the data
- this table has the CEL file location in the first column, without a header
- the other columns contain the sampleIDs, ages, genders,ethnicities and respective studies

- also limit samples to those wihtin the age ranges described above and label each with their "stage"

```{r covariatedataframe}
covartemplate.file <- "unified-metadata-refined_HT.tab" 
covartemplate <- read.table(file.path(metaDir,covartemplate.file ), header=T, colClasses="character", sep="\t")
# convert character ages to numeric to allow numeric subsetting 
covartemplate$age <- as.numeric(covartemplate$age) 
## limit samples to those within the described child, adult and neonate age ranges
covartemplate <- covartemplate[union(union(which(covartemplate$age<=max(child.age.range) & covartemplate$age>=min(child.age.range)), which(covartemplate$age<min(child.age.range))), which(covartemplate$age<=max(adult.age.range) & covartemplate$age>=min(adult.age.range))),] 
# convert age back to character value so you can use it in a character matrix
covartemplate$age <- as.character(covartemplate$age)
covartemplate <- as.matrix(covartemplate)
covars <- aaply(covartemplate, 1, function(n){
  # pull out the info you need piece by piece
  sampleID <- n[1] 
  age <- n[3]
  gender <- n[4]
  ethnicity=n[5]
  CELregex <- n[7]
  CELFTP <- n[6]
  study=n[9]
  platform=n[8]
  if(n[10]=="GEO"){
    # for GEO studies, get the CEL file name from the FTP locaiton in the metadata file
    CELfileloc <- file.path(dataDir,study, sub(".gz", "", basename(as.character(CELFTP)))) 
  } else {
    # for ArrayExpress studies, get the CEL file name directly from the CEL file regex in the metadata file
    CELfileloc <- file.path(dataDir, study, CELregex) 
    }
  if(!file.exists(CELfileloc)){
    # check if the CEL file is actually where its supposed to be, if not notify and move on
    print(paste(CELfileloc, "does not exist", sep=" ")) 
    break
    } else {
      # if CEL file is where its supposed to be, add location to covariate table
      return(list(CELfileloc=CELfileloc, ID=sampleID, age=age, gender=gender, ethnicity=ethnicity, study=study, platform=platform)) 
      }
  })
covars <- as.data.frame(covars)
covars$age <- as.numeric(covars$age)
# label samples with their stage as determined by age
# don't need to specify adult ranges, as we limited the dataset above to neonates, children and adults
covars$stage <- ifelse(covars$age<max(neonate.age.range), "neonate", ifelse(covars$age>=min(child.age.range) & covars$age<=max(child.age.range), "child", "adult")) 
```

### Load Illumina data files
- do this separately for each array type

```{r load_data, eval=FALSE}
GPL.groups <- list(HT3="GPL6947", HT4="GPL10558")
annotation.labels <- list(HT3="illuminaHumanv3", HT4="illuminaHumanv4")
for(n in names(GPL.groups)) {
  GPL <- GPL.groups[[n]]
  covars.array <- covars[covars$platform %in% GPL,]
  exprs.matrix <- laply(as.character(covars.array[,"CELfileloc"]), function(x){
    data.raw <- as.matrix(read.delim(as.character(unlist(x)), header=T, row.names=1))
    raw.exprs <- data.raw[,1] 
    return(raw.exprs)
    })
  se.exprs.matrix <- laply(as.character(covars.array[,"CELfileloc"]), function(x){
    data.raw <- as.matrix(read.delim(as.character(unlist(x)), header=T, row.names=1))
    raw.se.exprs <- data.raw[,2] 
    return(raw.se.exprs)
    })
  beadNum.matrix <- laply(as.character(covars.array[,"CELfileloc"]), function(x){
    data.raw <- as.matrix(read.delim(as.character(unlist(x)), header=T, row.names=1))
    if(ncol(data.raw)==3){
      raw.beadnums <- data.raw[,3] 
      } else {raw.beadnums=rep(NA, nrow(data.raw))}
    return(raw.beadnums)
    })
  exprs.matrix <- t(exprs.matrix) 
  dimnames(exprs.matrix)[[2]] <- covars.array$ID
  se.exprs.matrix <- t(se.exprs.matrix) 
  dimnames(se.exprs.matrix)[[2]] <- covars.array$ID
  beadNum.matrix <- t(beadNum.matrix) 
  dimnames(beadNum.matrix)[[2]] <- covars.array$ID
  if(all(is.na(beadNum.matrix))){
    lumibatch <-  new('LumiBatch', exprs=exprs.matrix, se.exprs=se.exprs.matrix)
    } else {
      lumibatch <-  new('LumiBatch', exprs=exprs.matrix, se.exprs=se.exprs.matrix, beadNum=beadNum.matrix)
      }
  row.names(covars.array) <- covars.array$ID
  pData(lumibatch) <- covars.array
  
  assign(paste("lumibatch", n, sep="."), lumibatch)
  rm(lumibatch)
  }
lapply(mic.raws, annotation) # confirm annotations
```

### QC checks of raw data

```{r QC_raw, eval=FALSE}
for(n in 1:length(GPL.groups)){
  lumibatch <- get(paste("lumibatch", names(GPL.groups)[n], sep="."))
  arrayQualityMetrics(expressionset=lumibatch, outdir=file.path(resultsDir, "QCreport_raw", names(GPL.groups)[n]), force=TRUE, do.logtransform=TRUE, intgroup=c("stage", "study"))
  rm(lumibatch)
  }
```

[HT3 Raw Data QC Report](./results/QCreport_raw/HT3/index.html)
[HT4 Raw Data QC Report](./results/QCreport_raw/HT4/index.html)


```{r virtualarray}
my_virtualArrays <- NULL
my_virtualArrays$noBatchEffect <- virtualArrayExpressionSets(all_expression_sets=TRUE,removeBatcheffect = F )

## Quality Control and Normalization

### Array QC checks on raw data with [ArrayQualityMetrics][3]

```{r rawQC, eval=FALSE}
arrayQualityMetrics(expressionset=mic.raw, outdir=file.path(resultsDir, "QCreport_raw"), force=TRUE, do.logtransform=TRUE, intgroup=c("stage", "study"))
````

[Raw Data QC Report](./results/QCreport_raw/index.html)

### Normalize all arrays using [RMA][4]

```{r normalize_RMA, eval=FALSE}
mic.norm.eset <- rma(mic.raw, normalize=TRUE, background=TRUE)
```

### Array QC checks on normalized data with ArrayQualityMetrics

```{r normQC, eval=FALSE}
arrayQualityMetrics(expressionset=mic.norm.eset, outdir=file.path(resultsDir, "QCreport_norm"), force=TRUE, do.logtransform=FALSE, intgroup=c("stage", "study"))
```

[Normalized Data QC Report](./results/QCreport_norm/index.html)

### Remove outlier studies
- largely based on array intensity and NUSE analyses of raw data
- removed studies where samples were still outliers after normalization
  - drop GSE18897, MEXP-2917, TABM-940 and TABM-666
- renormalize data

```{r drop_outliers, eval=FALSE, results='asis'}
mic.raw <- mic.raw[,which(!(unlist(pData(mic.raw)$study) %in% c("GSE18897", "MEXP-2917", "TABM-940", "TABM-666")))]
mic.norm.eset <- rma(mic.raw, normalize=TRUE, background=TRUE)
save.image(file.path(resultsDir, "RDATA.raw_and_normalized_microarray.data_U133Plus2.0" ))
```

[Normalized Data QC Report after removing outlier studies](./results/QCreport_norm.wo.outliers/index.html)

```{r load_affybatch, echo=FALSE}
# this is a hack to get around knitr having memory issues with loading and parsing CEL files
# instead of having knitr do the CEL file manipulation, do it interactively and save out an RDATA object that you can load into the final knitr run
load(file.path(resultsDir, "RDATA.raw_and_normalized_microarray.data_U133Plus2.0" ))
```

### Metadata for remaining U133Plus2.0 studies

```{r print_metadata, results='asis'}
pd <- pData(mic.norm.eset)
pd.gvis <- gvisTable(as.data.frame(apply(pd, 2, as.character)), options=list(width=960))  
print(pd.gvis, "chart")
```

### Repeat Array QC checks

```{r postoutlierQC, eval=FALSE}
arrayQualityMetrics(expressionset=mic.norm.eset, outdir=file.path(resultsDir, "QCreport_norm.wo.outliers"), force=TRUE, do.logtransform=FALSE, intgroup=c("stage", "study"))
```

## Differential Expression Analyses

### Limma

A linear model for microarray data analysis [Limma][5] was performed on the samples to identify differentially expressed genes between all pair-wise comparisons of the sample groups. Limma fits a linear model to the expression data for all samples for each gene and is designed to handle complex experiments involving comparisons between many RNA targets simultaneously.

To perform limma, we construct two matrices. The design matrix provides a representation of the different sample groups which have been analysed. The contrast matrix allows the coefficients defined by the design matrix to be combined into contrasts of interest. Each contrast corresponds to a pairwise comparison between sample groups.

#### Design matrix 
- make a matrix with arrays as rows, sample groups as columns
- a one or a zero indicate respectively, that a sample either belongs or does not belong to the sample group 

```{r design, results="asis"}
design <- model.matrix(~ -1+factor(pData(mic.norm.eset)$stage))
# make sure the headings match
colnames(design) <- sub("factor\\(pData\\(mic.norm.eset\\)\\$stage\\)", "", colnames(design))
design.gvis <- gvisTable(as.data.frame(apply(design, 2, as.character)), options=list(width=240))  
print(design.gvis, "chart")
```

#### Contrast matrix
- to perform specified pairwise comparisons
- in this table, columns are contrasts/comparisons and rows are sample groups
-  a zero denotes that the sample group is not involved in the contrast, a 1 denotes that it has higher expression in the contrast and a -1 denotes lower expression in the contrast

```{r contrastmatrix, results='asis'}
contrast.matrix <- makeContrasts(adult-child,adult-neonate,neonate-child, levels=dimnames(design)[[2]])
contrast.gvis <- gvisTable(as.data.frame(apply(contrast.matrix, 2, as.character)), options=list(width=240, height=120))  
print(contrast.gvis, "chart")
```

These matrices are used to fit a linear model to the data. The linear model is applied and pairwise comparisons are performed to identify differentially expressed genes.

- first fit the linear model based on the design matrix for each gene based on the given series of arrays
- using the contrast matrix, compute estimated coefficients and standard errors for contrasts
- compute moderated t-statistics and log-odds of differential expression by empirical Bayes shrinkage of the standard errors towards a common value

#### Fit a linear model
- for each gene based on the given series of arrays

```{r linearmodel}
exprs.norm <- exprs(mic.norm.eset)
dimnames(exprs.norm)[[2]] <- as.character(pData(mic.norm.eset)$ID)
fit.exprs <- lmFit(exprs.norm, design) 
```

#### Compute estimated coefficients and standard errors for contrasts

```{r contrastfit}
fit2.exprs <- contrasts.fit(fit.exprs, contrast.matrix) 
```

#### Compute moderated t-statistics and log-odds of differential expression 
- by empirical Bayes shrinkage of the standard errors towards a common value

```{r bayes}
fit2.exprs <- eBayes(fit2.exprs) 
```

### Obtain top results
- generate a list of top 100 differentially expressed genes with at least 2 fold changes in expression and an adjusted pvalue <0.1,  as sorted by unadjusted pvalue

#### The summary table contains the following information: 
- logFC is the log2-fold change
- the AveExpr is the average expression value accross all arrays
- the moderated t-statistic (t) is the logFC to its standard error, the P.Value is the associated p-value
- the adj.P.Value is the p-value adjusted for multiple testing (by FDR) 
- the B-value (B) is the log-odds that a gene is differentially expressed (the-higher-the-better)

```{r toptable, results='hide'}
top.stats <- lapply(seq(1,3,1), function(n) {
  contrast <- dimnames(fit2.exprs$contrasts)$Contrasts[n]
  stats.top <- topTable(fit2.exprs, coef=n, adjust="fdr", sort.by="p", p.value=0.1, lfc=1, number=100, genelist=row.names(exprs.norm)) 
  # expression levels for these genes in all samples
  if(nrow(stats.top)>0) {
    exprs.norm.sub  <-  exprs.norm[stats.top[, "ID"], ] 
    stats.exprs.norm.top <- cbind(stats.top, exprs.norm.sub)
    } else { stats.exprs.norm.top <-"NA"}
  return(list(data=stats.exprs.norm.top, contrast=contrast))
  })
# write stats and expression levels to file
lapply(top.stats, function(n) {
  contrast <- gsub(" ", "",  n$contrast)
  output <- n$data
  symbols <- unlist(mget(output$ID, hgu133plus2SYMBOL, ifnotfound=NA))
  output <- cbind(symbols, output)
    write.table(output, file.path(resultsDir, paste("U133_2.0Plus.top100.DE.probesets", contrast, "xls", sep=".")), sep="\t", row.names=F, col.names=T)
  })
```

#### Excel files with statistics and normalized sample intensities for the top 100 differentially expressed probesets for each comparison

[Adult vs. Child](./results/U133_2.0Plus.top100.DE.probesets.adult-child.xls)  
[Adult vs. Neonate](./results/U133_2.0Plus.top100.DE.probesets.adult-neonate.xls)  
[Neonate vs. Child](./results/U133_2.0Plus.top100.DE.probesets.neonate-child.xls)  

### Heatmaps of top 100 differentially expressed genes for each pairwise comparison
- there are three heatmaps for each comparison
  - both genders 
  - female only
  - male only

*note that the differentially expressed genes clustered in these heatmaps are based on comparisons that include both genders* 

```{r heatmaps, fig.width=24, fig.height=24}
for(n in 1:3){
  # get the names of the stages that are being compared in this comparison
  contrast <- top.stats[[n]]$contrast
  stages <- names(which(abs(contrast.matrix[,contrast])>0))
  # subset the samples and the data to those in the stages compared in this comparison
  sample.indices <- which(design[,stages[1]]>0 | design[,stages[2]]>0)
  sample.pd <- pData(mic.norm.eset)[sample.indices,]
  top.exprs.contrast <- top.stats[[n]]$data
  row.names(top.exprs.contrast) <- top.exprs.contrast$ID
  top.exprs.contrast <- top.exprs.contrast[,unlist(sample.pd$ID)]
  # setup row names for the heatmap, paste probeset ID to gene symbol
  row.names(top.exprs.contrast) <- paste(as.vector(unlist(mget(row.names(top.exprs.contrast), hgu133plus2SYMBOL, ifnotfound=NA))), " (" ,row.names(top.exprs.contrast), ")", sep="")
  # remove AFFY control probes
  if (any(grepl("AFFX", row.names(top.exprs.contrast)))){
    top.exprs.contrast <- top.exprs.contrast[-(grep("AFFX", row.names(top.exprs.contrast))),]
    }
  # heatmap annotations
  heatmap.annots <- sample.pd[,c("ID", "study", "stage", "gender")]
  heatmap.annots <- as.data.frame(apply(heatmap.annots, 2, unlist))
  row.names(heatmap.annots) <- heatmap.annots$ID
  heatmap.annots$ID <- NULL
  # heatmap annotation colors
  study_colors <- c("#FF0000","#00FF00", "#0000FF", cbPalette )
  names(study_colors) <- unique(unlist(pd$study))
  stage_colors <- c("white", "darkgrey", "black")
  names(stage_colors) <- unique(unlist(pd$stage))
  gender_colors <- c("cyan", "pink")
  names(gender_colors) <- unique(unlist(pd$gender))
  ann_colors = list(study = study_colors, stage = stage_colors, gender=gender_colors)
  ## Heatmaps
  # ALL genders 
  pheatmap(as.matrix(top.exprs.contrast), annotation=heatmap.annots, color=rev(brewer.pal(11,"RdBu")), main=contrast, show_colnames=F, fontsize=24, fontsize_row=10,annotation_colors=ann_colors)  
  # FEMALE gender  
  top.exprs.contrast.female <- top.exprs.contrast[,which(sample.pd$gender=="FEMALE")]
  pheatmap(as.matrix(top.exprs.contrast.female), annotation=subset(heatmap.annots,gender=="FEMALE"), color=rev(brewer.pal(11,"RdBu")), main=paste(contrast,"(FEMALE)", sep=" "), show_colnames=F,fontsize=24, fontsize_row=10,annotation_colors=ann_colors)  
  # MALE gender  
  top.exprs.contrast.male <- top.exprs.contrast[,which(sample.pd$gender=="MALE")]
  pheatmap(as.matrix(top.exprs.contrast.male), annotation=subset(heatmap.annots,gender="MALE"), color=rev(brewer.pal(11,"RdBu")), main=paste(contrast,"(MALE)", sep=" "), show_colnames=F, fontsize=24,fontsize_row=10,annotation_colors=ann_colors)  
  }
```

### Heatmaps for all top 100 differentially expressed genes from each pairwise comparison in all samples
- wanted to see if patterns were consistent across all the age groups
- combined the top hits from all three comparisons, and looked at their expression patterns in all samples

```{r heatmaps_all_samples, fig.width=24, fig.height=36}
top.IDs <- unique(unlist(lapply(top.stats, function(n) n$data$ID)))
top.exprs.union <- exprs.norm[top.IDs,]
# remove any AFFY control probes
if (any(grepl("AFFX", row.names(top.exprs.union)))){
  top.exprs.union <- top.exprs.union[-(grep("AFFX", row.names(top.exprs.union))),]
  }
# row labels - add gene symbol to probeset id
row.names(top.exprs.union) <- paste(as.vector(unlist(mget(row.names(top.exprs.union), hgu133plus2SYMBOL, ifnotfound=NA))), " (" ,row.names(top.exprs.union), ")", sep="")
# annotations
heatmap.annots <- pd[,c("ID", "study", "stage", "gender")]
heatmap.annots <- as.data.frame(apply(heatmap.annots, 2, unlist))
row.names(heatmap.annots) <- heatmap.annots$ID
heatmap.annots$ID <- NULL
# annotation colors
study_colors <- c("#FF0000","#00FF00", "#0000FF", cbPalette )
names(study_colors) <- unique(unlist(pd$study))
stage_colors <- c("white", "darkgrey", "black")
names(stage_colors) <- unique(unlist(pd$stage))
gender_colors <- c("cyan", "pink")
names(gender_colors) <- unique(unlist(pd$gender))
ann_colors = list(study = study_colors, stage = stage_colors, gender=gender_colors)
## Heatmaps
# Both genders
pheatmap(as.matrix(top.exprs.union), annotation=heatmap.annots, color=rev(brewer.pal(11,"RdBu")), main="All Comparisons, All Samples", show_colnames=F, fontsize=24,fontsize_row=8,annotation_colors=ann_colors)  
# Female gender
top.exprs.union.female <- top.exprs.union[,which(pd$gender=="FEMALE")]
pheatmap(as.matrix(top.exprs.union.female), annotation=subset(heatmap.annots,gender=="FEMALE"), color=rev(brewer.pal(11,"RdBu")), main="All Comparisons, All Female Samples", show_colnames=F,fontsize=24, fontsize_row=8,annotation_colors=ann_colors)  
# Male gender
top.exprs.union.male <- top.exprs.union[,which(pd$gender=="MALE")]
pheatmap(as.matrix(top.exprs.union.male), annotation=subset(heatmap.annots,gender=="MALE"), color=rev(brewer.pal(11,"RdBu")), main="All Comparisons, All male Samples", show_colnames=F, fontsize=24,fontsize_row=8,annotation_colors=ann_colors)   
```


```{r save, echo=FALSE}
# passing on data to the barcode analyses
rm(adult.age.range, ann_colors, child.age.range, contrast, contrast.gvis, contrast.matrix, covars, covartemplate.file, design, design.gvis, exprs.norm, fit2.exprs, fit.exprs, gender_colors, heatmap.annots, mapCdfName, n, neonate.age.range, pd, pd.gvis, sample.indices, metaDir, baseDir,  sample.pd, covartemplate, dataDir, stage_colors, stages, top.exprs.contrast, top.exprs.contrast.female, top.exprs.contrast.male, top.exprs.union, top.exprs.union.female, top.exprs.union.male, top.IDs, study_colors, cbPalette, age, CELfileloc, CELFTP, CELregex, covartemxplate.file, ethnicity, gender, refined.metadata, sampleID, study, temp, U133.Plus2.data)
save.image(file.path(resultsDir, "RDATA.raw_and_normalized_microarray.data.topstats.U133Plus2.0"))
```


[1]: http://www.bioconductor.org (BioC)
[2]: http://www.bioconductor.org/packages/release/bioc/html/affy.html (affy)
[3]: http://www.bioconductor.org/packages/2.10/bioc/html/arrayQualityMetrics.html (arrayQualityMetrics)
[4]: http://bioinformatics.oxfordjournals.org/content/19/2/185 (RMA)
[5]: http://www.bioconductor.org/packages/release/bioc/html/limma.html (limma)

