---
title: "Sahay Lab Klf9 differential expression analysis"
date: "`r BiocStyle::doc_date()`"
author: "Michael J. Steinbaugh"
bibliography: bibliography.bib
---

[Bioconductor]: https://bioconductor.org
[Ensembl]: http://useast.ensembl.org/Mus_musculus/Info/Index
[R]: https://www.r-project.org

[`DESeq2`]: https://bioconductor.org/packages/release/bioc/html/DESeq2.html
[`sailfish`]: http://www.cs.cmu.edu/~ckingsf/software/sailfish

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = TRUE,
                      echo = FALSE,
                      message = FALSE,
                      warning = FALSE)

# S4Vectors standardGeneric `rename()` masks `dplyr::rename()`
library(DESeq2)
library(DEGreport)
library(pheatmap)
library(tidyverse)

data(dds,
     dds_rlog,
     dds_vst,
     heatmap_annotation,
     metadata,
     normalized_counts,
     summary_data)
```

```{r ensembl}
library(biomaRt)
ensembl_version <- listMarts(host = "useast.ensembl.org") %>%
    dplyr::filter(biomart == "ENSEMBL_MART_ENSEMBL") %>%
    dplyr::select(version) %>% .[[1]]
mart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
mart_options <- listAttributes(mart)
metadata_ensembl <-
    getBM(mart = mart,
          attributes = c("ensembl_gene_id",
                         "external_gene_name",
                         "description",
                         "gene_biotype")) %>%
    arrange(ensembl_gene_id)
saveData(metadata_ensembl)
rm(mart)
detach("package:biomaRt", unload = TRUE)
```



# Preface

Differential gene expression (DGE) analysis of count data from [`sailfish`][] [@patro2014sailfish] was performed with the [Bioconductor][] [R][] package [`DESeq2`][] [@love2014deseq2]. Counts were fit to a negative binomial model [@hilbe2011negative], and dispersion estimates were generated using the mean values from the maximum likelihood estimate of log<sub>2</sub> fold changes, optimizing the Cox-Reid adjusted profile likelihood [@cox-reid1987].



# [`DESeq2`][] fit modeling

Several quality metrics were first assessed to explore the fit of the model, before differential expression analysis was performed. We observe that the modeling fit is good.

The plots below show the standard deviation of normalized counts (`normalized_counts`) using `log2()`, `DESeq2::rlog()`, and variance stabilizing (`DESeq2::vst()`) transformations by `rank(mean)`. The transformations greatly reduce the standard deviation, with `rlog` stabilizing the variance best across the mean. Therefore, we will use the `rlog` transformed counts for any downstream count visualizations.


## Variance stabilization

### `log2`

```{r deseq2_qc_log2}
library(vsn)
meanSdPlot(log2(normalized_counts + 1))
```

### `rlog` (used)

```{r deseq2_qc_rlog}
meanSdPlot(assay(dds_rlog))
```

### `vst`

```{r deseq2_qc_vst}
meanSdPlot(assay(dds_vst))
detach("package:vsn", unload = TRUE)
```


## Dispersion

The following plot shows the dispersion by mean of normalized counts. We expect the dispersion to decrease as the mean of normalized counts increases. This looks good.

```{r dispersion_estimates}
plotDispEsts(dds)
```



# Alpha level (a.k.a. false discovery rate) cutoffs

Let's take a look at the number of genes we get with different false discovery rate (FDR) cutoffs. These tests subset *P* values that have been multiple test corrected using the Benjamini Hochberg (BH) method [@benjamini-hochberg1995].


## `alpha` = 0.1 (default)

```{r}
summary(results(dds))
```


## `alpha` = 0.05

```{r}
summary(results(dds, alpha = 0.05))
```


## `alpha` = 0.01

```{r}
summary(results(dds, alpha = 0.01))
```


## `alpha` = 1e-6

```{r}
summary(results(dds, alpha = 1e-6))
```


Based on these results, it appears that an `alpha` cutoff of **0.01** is the sweet spot.

```{r res}
res <- results(dds, alpha = 0.01)
alpha <- res@metadata$alpha
saveData(res)
# mcols(res)
```

```{r expt_vs_control, include = FALSE}
expt_vs_control <- res %>%
    as.data.frame %>%
    rownames_to_column("ensembl_gene_id") %>%
    as_tibble %>%
    left_join(metadata_ensembl, by = "ensembl_gene_id") %>%
    setNamesSnake %>%
    arrange(padj)

expt_vs_control_de <- subset(expt_vs_control, padj < alpha)
expt_vs_control_de_dn <- subset(expt_vs_control_de, log2_fold_change < 0)
expt_vs_control_de_up <- subset(expt_vs_control_de, log2_fold_change > 0)

saveData(expt_vs_control,
         expt_vs_control_de,
         expt_vs_control_de_dn,
         expt_vs_control_de_up)
writeCsv(expt_vs_control,
         expt_vs_control_de,
         expt_vs_control_de_dn,
         expt_vs_control_de_up)

nrow(expt_vs_control)
nrow(expt_vs_control_de)
nrow(expt_vs_control_de_dn)
nrow(expt_vs_control_de_up)
```



# Summary

- `r nrow(expt_vs_control)` genes evaluated
- `r nrow(expt_vs_control_de)` genes differentially expressed
- `r nrow(expt_vs_control_de_up)` genes up-regulated
- `r nrow(expt_vs_control_de_dn)` genes down-regulated



# Plots

## Mean average (MA) plot

An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@yang2002ma].

```{r ma, fig.width = 7, fig.height = 7}
plotMA(res, ylim = c(-6, 6))
```


## Volcano plot

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log<sub>2</sub>) [@cui2003volcano; @li2014volcano]. We observe a few genes with very high changes in expression (LFC > 5).

```{r volcano}
library(CHBUtils)
library(ggrepel)
volcano_alpha <- 1e-3
volcano_lfc <- 1
volcano_lfc_outlier <- 10

# `volcano_density_plot()` requires a `data.frame` with two columns:
# `logFC` and `Adjusted.Pvalue`
# First show the high logFC outliers, then filter and re-graph.
volcano_df <- res %>%
    as.data.frame %>%
    dplyr::rename(Adjusted.Pvalue = padj,
                  logFC = log2FoldChange) %>%
    .[, c("logFC", "Adjusted.Pvalue")] %>%
    # Filter zero counts
    .[!is.na(.$logFC), ]

# https://github.com/hbc/CHBUtils/blob/master/R/volcanoPlot.R
volcano_plot_text <- volcano_df %>%
    rownames_to_column("ensembl_gene_id") %>%
    left_join(metadata_ensembl, by = "ensembl_gene_id") %>%
    mutate(name = external_gene_name) %>%
    select(logFC, Adjusted.Pvalue, name) %>%
    arrange(name) %>%
    subset(Adjusted.Pvalue < volcano_alpha &
               (logFC < -volcano_lfc | logFC > volcano_lfc))
```

```{r volcano_outlier, fig.width = 6, fig.height = 6}
volcano_plot_text_outlier <- subset(volcano_plot_text,
                                   logFC < -volcano_lfc_outlier |
                                       logFC > volcano_lfc_outlier)
volcano_density_plot(volcano_df,
                     lfc.cutoff = volcano_lfc_outlier,
                     plot_text = volcano_plot_text_outlier,
                     pval.cutoff = volcano_alpha,
                     shade.colour = "yellow",
                     title = "Volcano Plot: High Fold-Change Outliers")
```

Let's filter those outliers on our graph to have a clearer looks at the results. Now we can see **Klf9** on the volcano plot clearly. Genes in the green box with text labels have an LFC above `r volcano_lfc` and an adjusted *P* value below `r volcano_alpha`. These are likely the top candidate genes of interest.

```{r volcano_outlier_removed, fig.width = 8, fig.height = 8}
volcano_df_subset <- subset(volcano_df,
                            logFC > -volcano_lfc_outlier &
                                logFC < volcano_lfc_outlier)
volcano_plot_text_subset <- subset(volcano_plot_text,
                                   Adjusted.Pvalue < volcano_alpha &
                                       (logFC < -volcano_lfc | logFC > volcano_lfc))
volcano_density_plot(volcano_df_subset,
                     lfc.cutoff = volcano_lfc,
                     plot_text = volcano_plot_text_subset,
                     pval.cutoff = volcano_alpha)
detach("package:CHBUtils", unload = TRUE)
```


## Heatmap

This plot shows only differentially expressed genes (n = `r nrow(expt_vs_control_de)`) on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@ward1963clustering]. We can see that the samples cluster nicely across conditions.

```{r heatmap_de}
# Alternate inputs: tpm, normalized_counts
dds_rlog %>%
    assay %>%
    # DE genes only
    .[expt_vs_control_de$ensembl_gene_id, ] %>%
    pheatmap(annotation = heatmap_annotation,
             # color = heatmap_color,
             clustering_distance_cols = "correlation",
             clustering_method = "ward.D2",
             main = "Differentially expressed genes",
             scale = "row",
             show_rownames = FALSE)
```



# Tables

Only the top 50 up- and down-regulated genes (arranged by *P* value) are shown.

```{r expt_vs_control_de_up}
top_table <- function(df, caption) {
    df %>%
        dplyr::select(-c(description, lfc_se, pvalue, stat)) %>%
        head(n = 50) %>%
        kable(caption = caption)
}
top_table(expt_vs_control_de_up, "Top upregulated genes")
top_table(expt_vs_control_de_dn, "Top downregulated genes")
```



# File downloads

## Differential expression

- **[`expt_vs_control.csv`](../results/expt_vs_control.csv)**:
    All genes
- **[`expt_vs_control_de.csv`](../results/expt_vs_control_de.csv)**:
    Differentially expressed (DE) genes, both negative and positive
- **[`expt_vs_control_de_dn.csv`](../results/expt_vs_control_de_dn.csv)**:
    Downregulated DE genes
- **[`expt_vs_control_de_up.csv`](../results/expt_vs_control_de_up.csv)**:
    Upregulated DE genes

These files contain the following columns:

- `ensembl_gene_id`: [Ensembl][] gene identifier
- `base_mean`: The mean of the normalized counts for that gene for all samples
- `log2_fold_change`: log<sub>2</sub> fold change (MLE)
- `lfc_se`: log<sub>2</sub> standard error
- `stat`: Wald statistic
- `pvalue`: Walt test *P* value
- `padj`: BH adjusted Wald test *P* value (corrected for multiple comparisons; FDR)
- `external_gene_name`: [Ensembl][] name (a.k.a. symbol)
- `description`: [Ensembl][] description
- `gene_biotype`: [Ensembl][] biotype (e.g. `protein_coding`)


## Counts

We advise using only these files to assess and/or graph genes on an individual basis:

- **[`normalized_counts`](../results/normalized_counts.csv)**:
    [`DESeq2`][] normalized counts
- **[`tpm`](../results/tpm.csv)**:
    [`sailfish`][]-generated transcripts per million (TPM)

Raw counts should **only** be used to perform a new differential expression analysis. We can provide this file upon request. These counts will vary across samples due to differences in sequencing depth, and have not been normalized.



# Methods

Gene annotations were obtained from `r ensembl_version`.



# References
