---
title: "Correlation Analysis"
output:
  html_document:
    toc: true
    theme: united
    fig_width: 12
    fig_height: 12
    fig_caption: false
    highlight: textmate
---


```{r setup, echo=FALSE}
options(width=200)
library(knitr)
opts_chunk$set(tidy=TRUE, cache=FALSE, echo=FALSE, highlight=TRUE, figalign="center", warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120))

clientname="Andrew Gagne"
clientname2="Steve DePalma"
clientemail="agagne@genome.med.harvard.edu"
lablocation="Biopolymers Facility"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```
---

Correlation of bcbio.RNAseq results with previous results for `r clientname` (`r clientemail`), at `r lablocation` using data from the Seidman lab  (`r clientname2`).  Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

# METHODS SUMMARY


There are two sets of raw reads data to compare:

1) bcbio data: counts generated for individual genes (using the Ensembl GTF annotation) from the Star aligments
This dataset was generated with [bcbio][1], all reads were trimmed to remove adapter sequences and low quality calls.  Trimmed reads were aligned with [Star][2] and gene specific read counts for [Ensembl gene annotations][3] determined with the featureCounts program from the [subread][4] suite. 


2) previous data: counts generated for individual genes and samples from `r clientname2`

The sections below provide code to reproduce the post-counting results and plots. 

---

# SETUP

## Libraries

Bioconductor and R libraries used to process and visualize the data

```{r libraries}
library(plyr)
library(reshape2)
library(ggplot2)
library(biomaRt)
library(devtools)
install_git("git://github.com/hbc/CHBUtils.git") # misc personal utilities
library(CHBUtils)
library(knitr)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
```

## Variables

Working directories, cutoffs and other variables necessary to the analysis.

```{r variables}
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/seidman_rnaseq")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/seidnman_rnaseq"
  } else if (file.exists("/Users/johnhutchinson/projects/seidman_rnaseq")){
    baseDir <- "/Users/johnhutchinson/projects/seidman_rnaseq"
    } else {
      baseDir <- "/Volumes/ody/projects/seidman_rnaseq"
      }
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")
dataDir <- file.path(baseDir, "project/final/2014-04-24_project")
annotated.count.file <- file.path(dataDir, "combined.counts")

bcbio.counts.file <-file.path(dataDir, "annotated_combined.counts") 
prev.counts.file <- file.path(baseDir, "data/previous_results/LA_E14.5_LA_P0_LA_P10_LA_P3_LV_E14.5_LV_P0_LV_P10_LV_P3_RA_E14.5_RA_P0_RA_P10_RA_P3_RV_E14.5_RV_P0_RV_P10_RV_P3.expr")
```

---

# DATA IMPORT

```{r dataload_and_reshape, results='asis'}
bcbio.counts <- read.table(bcbio.counts.file, header=T, row.names=1)
prev.counts <- read.delim(prev.counts.file, sep="\t", header=T, row.names=1)
```

---

# DATA PREPROCESS

Subset to common samples and common gene symbols

```{r munge}
## subset previous data to raw counts and rename samples
prev.counts <- prev.counts[,grepl("reads", names(prev.counts))]
names(prev.counts) <- sub(".reads.", "", names(prev.counts))
names(prev.counts) <- c("E14_5_LA", "P0_LA"  ,  "P10_LA"  , "P3_LA"  ,  "E14_5_LV", "P0_LV",    "P10_LV",   "P3_LV"   , "E14_5_RA", "P0_RA"   , "P10_RA"  , "P3_RA"  ,  "E14_5_RV", "P0_RV" ,   "P10_RV"  , "P3_RV")

## subset bcbio data to genes with unique gene symbols
bcbio.counts <- bcbio.counts[order(bcbio.counts$symbol),]
bcbio.counts <- bcbio.counts[!duplicated(bcbio.counts$symbol),]

## now subset bcbio and previous data to common gene symbols
common.symbols <- intersect(bcbio.counts$symbol, row.names(prev.counts))
bcbio.counts <- bcbio.counts[bcbio.counts$symbol %in% common.symbols,]
prev.counts <- prev.counts[row.names(prev.counts) %in% common.symbols,]

## replace row.names of bcbio data
row.names(bcbio.counts) <- bcbio.counts$symbol
bcbio.counts$symbol <- NULL
# sort both datasets by gene symbol
bcbio.counts <- bcbio.counts[order(row.names(bcbio.counts)),]
prev.counts <- prev.counts[order(row.names(prev.counts)),]
bcbio.counts <- bcbio.counts[,order(names(bcbio.counts))]
prev.counts <- prev.counts[,order(names(prev.counts))]

bcbio.counts <- row2colnames(bcbio.counts, "symbol")
prev.counts <- row2colnames(prev.counts, "symbol")

# melt data and combine
bcbio.counts <- melt(bcbio.counts)
names(bcbio.counts) <- c("symbol", "sample", "bcbio_counts")
prev.counts <- melt(prev.counts)
names(prev.counts) <- c("symbol", "sample", "biopolymer_counts")
both.counts <- merge(bcbio.counts, prev.counts)
```

Correlations are based on `r length(common.symbols)` genes which were shared between the two dataset annotations.

Here I calculated the pearson correlation (r) of the gene counts for each sample.

```{r correlations, results='asis'}
by.sample.cor <- ddply(both.counts,"sample",function(x) cor(x$bcbio_counts,x$biopolymer_counts))
names(by.sample.cor) <- c("sample", "correlation")
kable(by.sample.cor)
```

You can look for systematic differences using MAplots.

```{r maplots}
ggplot(both.counts, aes(x=0.5*(log(bcbio_counts)+log(biopolymer_counts)), y=(log(biopolymer_counts)-log(bcbio_counts))))+
  geom_point(alpha=0.25)+
  facet_wrap(~sample, nrow=4)+
  scale_y_continuous(name="M (difference between log counts (biopolymer-bcbio))")+
  scale_x_continuous(name="A (average log counts)")
```

The previous method appears to find a larger number of counts for many genes than bcbio.

We can examine those genes individually to see if they are consistently the same genes.

```{r individualgenes}
both.counts$A <- 0.5*(log(both.counts$bcbio_counts)+log(both.counts$biopolymer_counts))
both.counts$M <- (log(both.counts$biopolymer_counts)-log(both.counts$bcbio_counts))





[1]: https://github.com/chapmanb/bcbio-nextgen (bcbio)
[2]: http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635 (Star)
[3]: http://tophat.cbcb.umd.edu/igenomes.html (iGenomes)
[4]: http://subread.sourceforge.net/ (subread)