---
bibliography: "references.bib"
csl: "bioinformatics.csl"
title: "Chung - Interneuron Stage  Analysis"
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_float: true
    toc_depth: 3
    number_sections: true
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library("knitcitations")
cleanbib()
options("citation_format" = "pandoc")

clientname="Sangmi Chung"
clientemail="sangmichung@gmail.com"
labPI="Sangmi Chung"
lablocation="McLean Hospital"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"

library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  


The purpose of this analysis is to find differentially expressed genes between:

- P40 interneurons and E13 cortex
- E13 cortex and E13 GE  

Will use microarray data from two separate studies:

- P40 interneurons are from  `r citep("10.1523/JNEUROSCI.0105-09.2009")` (GSE17806) 
- E12 samples  are from `r citep("10.1002/cne.22271")`

All MoGene 1.0 ST arrays (Faux et al.) were processed using the 'oligo' BioConductor package `r citep("10.1093/bioinformatics/btq431")` and all Mouse 430A2 arrays (Okaty et al.)  with the 'affy' Bioconductor package `r citep("10.1093/bioinformatics/btg405")`. 

All arrays were quality-controlled with arrayQualityMetrics `r citep("10.1093/bioinformatics/btn647")`, normalized with RMA `r citep("10.1093/biostatistics/4.2.249")`, filtered, subset to one probe per gene and further subset to common genes between teh two datasets. 

Genes most likely to be differentially expressed were identified by custom rank based means.

# Setup

### Variables
Working directories, files and other variables necessary to the analysis.

```{r variables}
# Setup Data and Results directory variables
if(file.exists("~/Work/projects/sc_interneuron_microarray")) {
  baseDir <- "~/Work/projects/sc_interneuron_microarray"
} else {
  baseDir <- "~/projects/sc_interneuron_microarray"
}
dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # colorblind friendly palette
covarsfilename="metadata.tab" # tab delimited file describing samples
lowintensity.percentile=0.1
mad.quantile.cutoff=0.1
pvalue.cutoff=0.05
highlight.color="green"
lfc.cutoff=1
numpcs=4
```

### Libraries

Bioconductor `r citep("10.1038/nmeth.3252")` and R `r citep("http://cran.r-project.org/")` libraries used to process the data.

```{r libraries, results='hide'}
library(oligo) # array utilities
library(affy)
library(arrayQualityMetrics) # array quality control reports
library(GEOquery)
library(WGCNA)
library(mogene10stprobeset.db) # array layout annotation
library(mogene10sttranscriptcluster.db) # array probe to gene annotations
library(mouse430a2.db)
library(clusterProfiler)
library(DOSE)

library(inSilicoDb)
library(inSilicoMerging)
library(RankProd)

library(ggdendro)
library(CHBUtils)
library(limma)
library(ggplot2)
library(pheatmap) # pretty heatmaps

library(readr)
library(dplyr)
library(tidyr)
library(rio)
library(DT)
library(janitor)

library(biomaRt)

# use older version of biomart to allow homology changes
mousemart <- useDataset("mmusculus_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL", host="useast.ensembl.org"))

mouseattributes = listAttributes(mousemart)
mousefilters = listFilters(mousemart)
```

---

# Import Data and Metadata {.tabset}

## Faux Data

- load in phenotypes and array names from metadata file for Faux data [metadata file](../meta/faux/metadata.tsv)
  - this file contains the names and descriptions of CEL files contained in the data directory 
- used these file names to load in the CEL files using the oligo package

```{r dataload.faux, results='hide'}
# simple tab delimited file with CEL file in first column  and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns
covars <- import(file.path(metaDir, "faux", "metadata.tsv")) %>%
  tbl_df() %>% 
  unite(., sampleid, age, tissue, replicate, sep="_", remove=FALSE) %>%
  mutate(., study="faux") %>%
  unite(., sampletype, age, tissue, remove=FALSE) %>%
  as.data.frame()
row.names(covars) <- covars$sampleid
setwd(file.path(dataDir, "faux"))
affyRaw.faux <- read.celfiles(as.character(covars$filename))
pData(affyRaw.faux) <- covars 
sampleNames(affyRaw.faux) <- pData(affyRaw.faux)$sampleid
validObject(affyRaw.faux)
rm(covars)
setwd(baseDir)
```

### Faux sample metadata

```{r covars.faux, results='asis'}
# Sample information table
kable(pData(affyRaw.faux), caption = "Faux dataset metadata", row.names=FALSE)
```

## Okaty Data

Here I pulled down the metadata and data from GEO using the GEOquery package `r citep("10.1093/bioinformatics/btm254")`. Data will be loaded in using the affy package.

```{r download.okaty.CELfiles}
getGEOSuppFiles("GSE17806", makeDirectory = FALSE, baseDir=file.path(dataDir, "okaty"))
# decompress data
setwd(file.path(dataDir, "okaty", "GSE17806"))
system("tar -xvf *.tar")
system ("gzip -d *.gz")
```

```{r dataload.okaty, results='hide'}
gse <- getGEO("GSE17806", getGPL=FALSE)
covars <- pData(gse[[1]]) %>%  tbl_df() %>%
  dplyr::select(.,geo_accession, characteristics_ch1, supplementary_file) %>%
  mutate(., filename=gsub(".*/", "", supplementary_file)) %>%
  mutate(., filename=sub(".gz$", "", filename)) %>%
  mutate(.,age=sub("age: ", "", characteristics_ch1)) %>% 
  mutate(., tissue="cortex", study="okaty") %>%
  mutate(., temp=1) %>% group_by(., age) %>% mutate(., replicate=cumsum(temp)) %>%
  unite(., sampleid, age, tissue, replicate, sep="_", remove=FALSE) %>%
  mutate(., study="okaty") %>%
  dplyr::select(., filename, sampleid, age, tissue, replicate,study) %>%
  unite(., sampletype,age, tissue, remove=FALSE) %>%
  as.data.frame()
row.names(covars) <- covars$sampleid
export(covars,file = file.path(metaDir, "okaty", "metadata.tsv"))

# load in cel files
celfiles <- as.character(unlist(covars$filename))
setwd(file.path(dataDir, "okaty", "GSE17806"))
affyRaw.okaty <- read.affybatch(filenames=celfiles)
pData(affyRaw.okaty) <- covars
sampleNames(affyRaw.okaty) <- covars$sampleid
validObject(affyRaw.okaty)
rm(covars)
setwd(baseDir)

# only care about the P40 samples, drop everything else

affyRaw.okaty <- affyRaw.okaty[,which(pData(affyRaw.okaty)$age=="P40")]
```

### Okaty sample metadata

```{r covars.okaty, results='asis'}
# Sample information table
kable(pData(affyRaw.okaty), caption = "Okaty dataset metadata", row.names=FALSE)
```

---

# Raw Data Qualilty Control {.tabset}

## Faux Data

### Raw Data QC

- using arrayQualityMetrics library `r citep("10.1093/bioinformatics/btn647")`

```{r rawQC.faux}
affyRaw.eset <- ExpressionSet(assayData=exprs(affyRaw.faux))
pData(affyRaw.eset) <- pData(affyRaw.faux)
data.prepped <- prepdata(affyRaw.eset, do.logtransform = TRUE, intgroup=c("age","tissue"))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "faux", "report_raw"), arrayTable = pData(affyRaw.eset), reporttitle = "Raw QC Report - Faux Data ")
```

The arrays all look good.

**[Raw Data QC Report - Faux Data](../results/faux/report_raw/index.html)**

## Okaty Data

### Raw Data QC

- using arrayQualityMetrics library `r citep("10.1093/bioinformatics/btn647")`

```{r rawQC.okaty}
affyRaw.eset <- ExpressionSet(assayData=exprs(affyRaw.okaty))
pData(affyRaw.eset) <- pData(affyRaw.okaty)
data.prepped <- prepdata(affyRaw.eset, do.logtransform = TRUE, intgroup=c("age","tissue"))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "okaty", "report_raw"), arrayTable = pData(affyRaw.eset), reporttitle = "Raw QC Report - Okaty Data ")
```

There is some variation in intensity distributions between the samples, but nothing terribly out of normal range.

**[Raw Data QC Report - Okaty Data](../results/okaty/report_raw/index.html)**

---

# Normalize Arrays (within dataset) {.tabset}

## Faux Data

### Background Correct and Normalize

- using RMA `r citep("10.1093/biostatistics/4.2.249")`

- summarize probesets on the gene ('core') level

```{r normalize.faux, results='hide'}
affyNorm.core.faux <- oligo::rma(affyRaw.faux,  target="core", background=TRUE, normalize=TRUE)
```

```{r normQC.faux}
data.prepped <- prepdata(affyNorm.core.faux, do.logtransform = FALSE, intgroup=c("tissue","age"))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "faux", "report_rma_core"), arrayTable = pData(affyNorm.core.faux), reporttitle = "Normalized QC Report - Faux Data")
```

The data now looks really good, good separation between the cortex and GE samples.

**[Normalized Data QC Report - Faux Data](../results/faux/report_rma_core/index.html)**

## Okaty Data

### Background Correct and Normalize

- using RMA `r citep("10.1093/biostatistics/4.2.249")`

```{r normalize.okaty, results='hide'}
affyNorm.okaty <- affy::rma(affyRaw.okaty,   background=TRUE, normalize=TRUE)
```

```{r normQC.okaty}
data.prepped <- prepdata(affyNorm.okaty, do.logtransform = FALSE, intgroup=c("tissue","age"))

aqmheat <- aqm.heatmap(data.prepped)
aqmbox <- aqm.boxplot(data.prepped)
aqmdens <- aqm.density(data.prepped)
aqmma <- aqm.maplot(data.prepped)

aqm.writereport(list(aqmheat,  aqmbox, aqmdens, aqmma), outdir = file.path(resultsDir, "okaty", "report_rma"), arrayTable = pData(affyNorm.okaty), reporttitle = "Normalized QC Report - Okaty Data")
```

The one array is still an outlier but will leave in for now as there are so few replicates.

**[Normalized Data QC Report - Okaty Data](../results/okaty/report_rma/index.html)**

--- 

# Combining Datasets

- annotate probes with gene names, collapse to single probe/measurement per gene and combine array platforms on common genes assayed

## Annotate Datasets {.tabset}

So far we have only been working with the probesets,without reference to the genes they assay. Here we load in metadata about the probesets on the array (feature data), the gene symbols and Entrez Gene IDs in particular.

### Faux Data

```{r features_faux}
featureData(affyNorm.core.faux) <- getNetAffx(affyNorm.core.faux, "transcript") # this will load the Affymetrix annotation, including the probeID, into the fData
# get gene symbols and entrezIDs for all probesets
fData(affyNorm.core.faux)$symbol <- as.character(unlist(mget(featureNames(affyNorm.core.faux), mogene10sttranscriptclusterSYMBOL, ifnotfound=NA))) # curated annotations from Bioconductor 
fData(affyNorm.core.faux)$entrezid <- as.character(unlist(mget(featureNames(affyNorm.core.faux), mogene10sttranscriptclusterENTREZID, ifnotfound=NA))) # curated annotations from Bioconductor 
kable(head(fData(affyNorm.core.faux)), caption = "Subset of array annotations", row.names=FALSE)
```

### Okaty Data

```{r features_okaty}
entrezids <- unlist(mget(featureNames(affyNorm.okaty), mouse430a2ENTREZID, ifnotfound=NA))
symbols <- unlist(mget(featureNames(affyNorm.okaty), mouse430a2SYMBOL, ifnotfound=NA))
probeids <- featureNames(affyNorm.okaty)
annots <- do.call(cbind, list(probeids, entrezids, symbols)) %>% as.data.frame()
names(annots) <- c("probeid", "entrezid", "symbol")
fData(affyNorm.okaty) <- annots
kable(head(fData(affyNorm.okaty)),caption = "Subset of array annotations", row.names=FALSE)

```

## Pre-Filter and Summarize Data {.tabset}

Dropping unwanted probes and collapsing to a single probe per gene before combining datasets.

### Faux Data

#### By Annotation
- subset to "main" category probes i.e. probes that assay genes
  - will also remove the control probes
- remove anything that doesn't have an annotated entrezID

```{r filter1_faux}
affyNorm.core.faux <- affyNorm.core.faux[which(fData(affyNorm.core.faux)$category=="main"),]
affyNorm.core.faux <- affyNorm.core.faux[which(!is.na(fData(affyNorm.core.faux)$entrezid)),]
```

`r nrow(affyNorm.core.faux)` probes remaining

#### By Cross Hybridization
- some probes are annotated as potentially hybridizing to multiple targets

```{r filter2}
affyNorm.core.faux <- affyNorm.core.faux[which(fData(affyNorm.core.faux)$crosshybtype=="1"),]
```

`r nrow(affyNorm.core.faux)` probes remaining

#### Collapse probes
- the array has multiple probes per gene, here I'll collapse this information to have on data point per gene using methods from the collapseRows function `r citep("10.1186/1471-2105-12-322")` in the WGCNA package.
 - here I selected the probeset for each gene with with the maximum mean value for all samples

```{r WGCNA_faux, results='hide'}
eset.core.annot <- cbind(exprs(affyNorm.core.faux), fData(affyNorm.core.faux)) %>%  tbl_df() %>% dplyr::select(., contains("E13"), entrezid, transcriptclusterid)
eset.core <- dplyr::select(eset.core.annot, contains("E13")) %>% as.data.frame()
row.names(eset.core) <- eset.core.annot$transcriptclusterid
collapsed.data <- collapseRows(eset.core,rowID=eset.core.annot$transcriptclusterid, rowGroup=eset.core.annot$entrezid , method="MaxMean")
affyNorm.core.faux <- affyNorm.core.faux[which(collapsed.data$selectedRow),]

fData(affyNorm.core.faux) <- fData(affyNorm.core.faux)[,c("symbol", "entrezid")]
row.names(fData(affyNorm.core.faux)) <- fData(affyNorm.core.faux)$entrezid
featureNames(affyNorm.core.faux) <- fData(affyNorm.core.faux)$entrezid
kable(head(fData(affyNorm.core.faux)),caption = "Subset of collapsed array annotations", row.names=FALSE)
```

`r nrow(affyNorm.core.faux)` genes remaining

### Okaty Data

#### By Annotation
- remove anything that doesn't have an annotated entrezID

```{r filter1_okaty}
affyNorm.okaty <- affyNorm.okaty[which(!is.na(fData(affyNorm.okaty)$entrezid)),]
```

`r nrow(affyNorm.okaty)` probes remaining

#### Collapse probes
- the array has multiple probes per gene, here I'll collapse this information to have on data point per gene using methods from the collapseRows function `r citep("10.1186/1471-2105-12-322")` in the WGCNA package.
 - here I selected the probeset for each gene with with the maximum mean value for all samples

```{r WGCNA_okaty, results='hide'}
eset.core.annot <- cbind(exprs(affyNorm.okaty), fData(affyNorm.okaty)) %>%  tbl_df() %>% dplyr::select(., contains("P40"), entrezid, probeid)
eset.core <- dplyr::select(eset.core.annot, contains("P40")) %>% as.data.frame()
row.names(eset.core) <- eset.core.annot$probeid
collapsed.data <- collapseRows(eset.core,rowID=eset.core.annot$probeid, rowGroup=eset.core.annot$entrezid , method="MaxMean")
affyNorm.okaty <- affyNorm.okaty[which(collapsed.data$selectedRow),]

fData(affyNorm.okaty) <- fData(affyNorm.okaty)[,c("symbol", "entrezid")]
row.names(fData(affyNorm.okaty)) <- fData(affyNorm.okaty)$entrezid
featureNames(affyNorm.okaty) <- fData(affyNorm.okaty)$entrezid
kable(head(fData(affyNorm.okaty)), caption = "Subset of collapsed array annotations", row.names=FALSE)
```

`r nrow(affyNorm.okaty)` genes remaining

## Merge Datasets

Using the insilicoDB Bioconductor package `r citep("10.1186/1471-2105-13-335")`.

```{r insilicomerge}
affynorm.merged <- merge(esets=list(affyNorm.core.faux, affyNorm.okaty),method='NONE')
kable(head(fData(affynorm.merged)), caption = "Subset of merged array annotations", row.names=FALSE)
```


### QC measures on merged data, without re-normalization

- these are the same plots as found in the QC reports on raw and normalized data above


```{r QC.merged}
data.prepped <- prepdata(affynorm.merged, do.logtransform = FALSE, intgroup=c("age","tissue", "study"))

aqm.heatmap(data.prepped)@plot
aqm.boxplot(data.prepped)@plot
aqm.density(data.prepped)@plot
aqm.maplot(data.prepped)@plot
```

---

# Functional Annotation

Used results from the Animal Transcription Factor Database `r citep(" 10.1093/nar/gku887")`. 

```{r GO_annots, results="hide"}
TFannots <- import(file.path(metaDir, "Mus_musculus_transcription_factors_gene_list.txt"), format="tsv") %>% clean_names() %>% tbl_df() %>% dplyr::select(.,-protein_have_dbd, -ensembl_id)
TFannots <- distinct(TFannots) # remove any duplicated annotations

annots <- fData(affynorm.merged) %>% tbl_df()
annots <- dplyr::left_join(annots,TFannots , by=c("entrezid"="entrez_id")) 
annots <-   mutate(annots, tf=ifelse(is.na(tf_family), "no", "yes")) 
annots <-   dplyr::select(annots, symbol, entrezid, tf, tf_family) %>% as.data.frame()
row.names(annots) <- annots$entrezid
identical(row.names(annots),featureNames(affynorm.merged))

fData(affynorm.merged) <- annots
validObject(affynorm.merged)
```

```{r showannots}
output1 <- fData(affynorm.merged) %>% tbl_df() %>% filter(., tf=="yes")
output <- rbind(head( fData(affynorm.merged)), head(output1))
kable(output, caption = "Subset of merged array annotations, with TF annotations", row.names=FALSE)
```

---

# Rank-based Analysis

- drop the E13 GE samples
- transform expression intensities into ranks for each sample and compare mean squared ranks for E14 and P40 samples

```{r ranking}
affynorm.merged <- affynorm.merged[,which(pData(affynorm.merged)$tissue=="cortex")]

eset <- exprs(affynorm.merged) %>%   
  tbl_df() %>%
  mutate(., entrezid=fData(affynorm.merged)$entrezid) %>%
  mutate(., symbol=fData(affynorm.merged)$symbol) %>%
  mutate(., tf=fData(affynorm.merged)$tf) %>%
  mutate(., tf_family=fData(affynorm.merged)$tf_family) %>%
  dplyr::select(., symbol, entrezid, tf, tf_family, contains("cortex"))

eset.ranked <-   mutate_each(eset, funs(ranked=min_rank), -entrezid, -symbol, -tf, -tf_family) %>%
  mutate_each(., funs(squared = (.)^2), contains("ranked")) %>%
  mutate(., E13_mean_squared_rank=dplyr::select(., contains("E13_cortex")) %>% 
           dplyr::select(., ends_with("squared")) %>% 
           rowMeans %>% 
           sqrt(.) %>%
           round(.,digits=0)) %>% 
  mutate(., P40_mean_squared_rank=dplyr::select(., contains("P40_cortex")) %>% 
           dplyr::select(., ends_with("squared")) %>% 
           rowMeans %>% 
           sqrt(.) %>% 
           round(.,digits=0)) %>%
  mutate(., rank_difference=P40_mean_squared_rank-E13_mean_squared_rank) %>%
  mutate(., ranked_rank_difference=percent_rank(rank_difference) ) %>%
  arrange(., desc(ranked_rank_difference, )) %>%
  dplyr::select(-contains("ranked_squared"))

export(eset.ranked,file.path(resultsDir, "ranked.results.xlsx"))

kable(head(eset.ranked,n=10), caption="Subset of ranked results, sorted by rank", row.names=FALSE)
```

----

# Downloads

[Ranks and expression levels of all genes for the comparison](../results/ranked.results.xlsx)

**The table contains the following information:**

- symbol - gene symbol
- entrezid - Entrez Gene ID
- _cortex - six columns of log normalized array intensities for each sample
- _ranked - six columns of gene ranks for each sample
    - **the higher the rank, the more the gene is expressed, so rank=1 means its the lowest expressed gene**
- _mean_squared_rank - 2 columns of mean of the squared ranks for each sample type (E13 & P40)
- rank_difference - the difference between the mean sequared ranks for P40 and E13, 
    - **a large value means the gene is higher expressed in P40 than in E13**
- ranked_rank_difference - the differences ranked for easy comparison from 0 to 1
    - **the closer the value to 1, the gene is higher expressed in P40 than in E13**

---

# References

```{r writebib, results='hide', echo=FALSE, message=FALSE, cache=FALSE}
write.bibtex(file="references.bib")
```