---
output:
  html_document:
    toc: true
    toc_depth: 3
    theme: united
title: "Splicing Microarray Analysis of TPRM6 Knockout Samples"
bibliography: "references.bib"
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitcitations)
cleanbib()
options("citation_format" = "pandoc")

arraytype="GeneChipÂ® Human Transcriptome Array 2.0"
clientname="Nora Renthal"
clientemail="Nora.Renthal@childrens.harvard.edu"
labPI="Clapham"
lablocation="Children's Hospital"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
sponsor="Harvard NeuroDiscovery Center"

library(knitr)
opts_chunk$set(warning=FALSE, error=FALSE, message=FALSE, echo=FALSE, cache=TRUE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev="svg")
options(width=200)
```

---

Array analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  

This analysis was subsidized by `r sponsor`.

Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

The sections below provide code to reproduce the included results and plots. 

---

# Methods Summary  

All `arraytype` arrays were processed using the 'oligo' BioConductor package `r citep("10.1093/bioinformatics/btq431")`, quality-controlled with arrayQualityMetrics `r citep("10.1093/bioinformatics/btn647")` and normalized with RMA `r citep("10.1093/biostatistics/4.2.249")`. Differentially expressed genes were identified using limma `r citep("http://link.springer.com/chapter/10.1007%2F0-387-29362-0_23")`.

---

# Setup

## Libraries

[Bioconductor](http://www.bioconductor.org) and [R](http://cran.r-project.org/) libraries used to process and visualize the data.

```{r libraries_variables}
library(oligo) # array utilities
library(pd.hta.2.0)# array layout annotation
library(hta20stprobeset.db)
library(hta20sttranscriptcluster.db)
library(arrayQualityMetrics) # array quality control reports
library(limma) # array statistical analyses
library(pheatmap) # pretty heatmaps
library(devtools) # install libraries from github
install_git("git://github.com/hbc/CHBUtils.git") # misc personal utilities
library(CHBUtils) # some homegrown functions
library(ggplot2) # pretty graphs
library(ggdendro) # for pretty dendrograms
library(biomaRt) # large biological database
library(dplyr)
```


## Variables
Working directories, files and other variables necessary to the analysis.

```{r variables}
## Setup Data and Results directory variables
if(file.exists("/n/hsphS10/hsphfs1/chb/projects/dc_splicingarray/")){
  baseDir="/n/hsphS10/hsphfs1/chb/projects/dc_splicingarray"
    } else if (file.exists("/Users/johnhutchinson/Work/projects/dc_splicingarray/")){
    baseDir="/Users/johnhutchinson/Work/projects/dc_splicingarray"
    } else {
      print("Can't find base directory")
    }

dataDir <- file.path(baseDir, "data")
metaDir <- file.path(baseDir, "meta")
resultsDir <- file.path(baseDir, "results")

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # colorblind friendly palette
covarsfilename="covars.csv" # tab delimited file describing samples

ensemblmart <- useMart("ENSEMBL_MART_ENSEMBL", dataset="hsapiens_gene_ensembl", host="www.ensembl.org")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)

lowintensity.percentile=0.1
mad.quantile.cutoff=0.1
pvalue.cutoff=0.05
highlight.color="green"
lfc.cutoff=log2(2)
```


## Functions

```{r functions}
# for plotting amount of variation explained by principal components
PCAplot.sd.eset <- function(eset=NULL,  title=NULL){
  eset.core <- exprs(eset)
  myPca.core <- prcomp(t(eset.core))
  # SD of components
  sdevdf <- data.frame(cbind(as.numeric(myPca.core$sdev),c(1:length(myPca.core$sdev))))
  sdevdf$prop <-  sdevdf$X1/sum(sdevdf$X1)
  sdevdf$cum <- cumsum(sdevdf$prop)
  ggplot(sdevdf, aes(x=X2, y=prop)) + 
    geom_point(size=4, color="red") + 
    scale_x_continuous('Component') + 
    scale_y_continuous('Standard Deviation') +
    ggtitle(title) +
    geom_line(data=sdevdf, aes(x=X2, y=cum))
}

# used for formatting labels on ggplots
fmt <- function(){ 
  function(x) format(x,nsmall = 1,scientific = FALSE)
}

QCarray <- function(eset, intgroup=NULL, outdir=NULL, logtransform=TRUE, title=NULL){
  
  prepdata <- prepdata(expressionset = eset, intgroup=intgroup, do.logtransform = logtransform)
  he <- aqm.heatmap(prepdata)
  bo <- aqm.boxplot(prepdata)
  de <- aqm.density(prepdata)
  ma <- aqm.maplot(prepdata)
  qm <- list("Boxplot"= bo, "Density"=de, "Heatmap"=he, "MAplot"=ma)
  aqm.writereport(modules=qm, reporttitle = title, outdir=outdir, arrayTable = pData(affyRaw))
}
```

---

# Import Data and Metadata

## Data

- load in phenotypes and array names from metadata file (`r covarsfilename`) in "meta" directory
  - this file contains the names and descriptions of CEL files contained in the data directory 
- use array names to load in arrays 

```{r dataload, results='hide'}
covars <- read.csv(file.path(metaDir, covarsfilename), row.names=1) # simple tab delimited file with CEL file in first column (no heading for this column) and sample metadata (i.e. sampleID, treatment group, batch etc.) in subsequent columns

celFiles <- file.path(dataDir, row.names(covars))
affyRaw <- read.celfiles(celFiles)
pData(affyRaw) <- covars 
sampleNames(affyRaw) <- pData(affyRaw)$sampleID
validObject(affyRaw)
```

## Sample metadata

```{r covars, results='asis'}
# Sample information table
kable(pData(affyRaw))
```

---

# PreProcessing 

## Raw Data 

### Quality Control

- using arrayQualityMetrics library `r citep("Kauffmann_2008")`

```{r rawQC, eval=FALSE}
QCarray(eset = affyRaw, intgroup = "genotype", outdir = file.path(resultsDir, "report_raw"), logtransform = TRUE, title = "Raw data QC")
```

[Raw Data QC Report](../results/report_raw/index.html)

The arrays look OK in terms of intensity distributions etc. However, they don't seem to be clustering well before normalization and one KO sample appears to be an outlier.

Normalizing the arrays may help clear this up.

## RMA Normalized Data

- background correct and normalize data with RMA `r citep("10.1093/bioinformatics/19.2.185")`

**Here I will summarize the probesets on both the the exon ('probeset') and gene ('core') levels.** THis is necessary to later adjust for overall gene expression levels when detecting differential exon expression.

```{r normalize, results='hide'}
affyNorm.gene <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
affyNorm.exon <- rma(affyRaw, target="probeset", background=TRUE, normalize=TRUE)
```

### Quality Control
- using arrayQualityMetrics library

```{r normQC, eval=FALSE}
QCarray(affyNorm.gene, intgroup="genotype", outdir=file.path(resultsDir, "report_norm.gene"), title="Normalized Gene level QC")
dev.off()
QCarray(affyNorm.exon, intgroup="genotype", outdir=file.path(resultsDir, "report_norm.exon"), title="Normalized Exon level QC")
dev.off()
```

[Normalized Data QC Report - Exon level](../results/report_norm.exon/index.html)
[Normalized Data QC Report - Gene level](../results/report_norm.gene/index.html)

The arrays look good with the exception of KO1, which is a clear outlier. I dropped KO1, and repeated the normalization.

```{r dropoutlier}
affyRaw <- affyRaw[,-which(pData(affyRaw)$sampleID=="KO1")]
affyNorm.gene <- rma(affyRaw, target="core", background=TRUE, normalize=TRUE)
affyNorm.exon <- rma(affyRaw, target="probeset", background=TRUE, normalize=TRUE)
```

```{r normQC2, eval=FALSE}
QCarray(affyNorm.gene, intgroup="genotype", outdir=file.path(resultsDir, "report_norm.gene_nooutlier"), title="Normalized Gene level QC")
dev.off()
QCarray(affyNorm.exon, intgroup="genotype", outdir=file.path(resultsDir, "report_norm.exon_nooutlier"), title="Normalized Exon level QC")
dev.off()
```

[Normalized Data QC Report without outlier sample - Exon level](../results/report_norm.exon_nooutlier/index.html)
[Normalized Data QC Report without outlier sample - Gene level](../results/report_norm.gene_nooutlier/index.html)


### Unsupervised Clustering of RMA Normalized Data

#### Hierarchical Clustering
The goal of these analyses are to naiively evaluate the variability within the raw data and determine whether this variability can predict the different sample groups

The first method produces a dendrogram by performing  
>  a hierarchical cluster analysis using a set of dissimilarities for the n objects being clustered

```{r cluster1, out.width='100%'}
plot_dendro(affyNorm.gene, labels.colname="genotype", colors.colname="genotype", title="Gene Level")
plot_dendro(affyNorm.exon, labels.colname="genotype", colors.colname="genotype", title="Exon level")
```

#### Principal Component Analysis (PCA)

This second approach is a dimension reduction and visualisation technique that is used to project the multivariate (i.e.multiple genes) data vector of each array into a lower-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the arrays. The data is typically reduced to a small number of dimensions (or components) which explain most of the sample variability. This [Youtube slideshow](https://www.youtube.com/watch?v=BfTMmoDFXyE) gives a pretty good basic explanation of what PCA is doing.

```{r PCAsd1, out.width='75%'}
PCAplot.sd.eset(affyNorm.gene, title="Gene level")
PCAplot.sd.eset(affyNorm.exon, title="Exon level")
```

Here, each point depicts the amount of variation explained by each component and the line shows the cumulative amount. For this data set, at both the gene and exon levels, the first 3 components explain just over 75% of the of the variation observed in the samples.

As plots with more than 2 dimensions are difficult to visualize, we typically  split up the dimensions/components and plot them pairwise against each other; the plots here show scatterplots of the arrays along all dual combinations of the first three principal components. 
You can use these plots to explore if the arrays cluster, find outliers, and determine whether this is according to an intended experimental factor or according to unintended causes such as batch effects. 

```{r pca1, out.width='100%'}
PCAplot.eset(affyNorm.gene, categories="genotype", title="Gene level", colorpalette=cbPalette, numcomponents=3, alpha=0.75)
PCAplot.eset(affyNorm.exon, categories="genotype", title="", colorpalette=cbPalette, numcomponents=3, alpha=0.75)
```

There is  a decent degree of clustering by sample type, with PC1 explaining much of the variation between the WT and KO samples.


----

# R Session Info

(useful if replicating these results)

```{r sessioninfo}
sessionInfo()
```

---

# References

```{r writebib, results='hide', echo=FALSE, message=FALSE}
write.bibtex(file="references.bib")
```

