---
output:
  knitrBootstrap::bootstrap_document:
    title: "Test file"
    theme: amelia
    highlight: sunburst
    theme.chooser: TRUE
    highlight.chooser: TRUE
---


# Load libraries
```{r libraries}
library(biomaRt)
library(knitr)
library(stringr)
library(xlsx)
library(ggplot2)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

```

# Functions
```{r functions}
bedTools.2in<-function(functionstring="bedtools intersect",bed1,bed2,opt.string="") {
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
 
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
 
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
 
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
bedTools.slop <- function(functionstring="bedtools slop", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  # create the command string and call the command using system()
  command <- paste(functionstring,"-i",a.file,"-g", genomefile, opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

bedTools.flank <- function(functionstring="bedtools flank -s", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out  
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  
  # create the command string and call the command using system()
  command <- paste(functionstring,"-g", genomefile, opt.string, "-i", a.file, ">", out, sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

```


# Set directories

```{r directories}
baseDir <- "~/projects/ab_nucmito/"
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
```

# Load data
- using mitochondrial data from the [microproteome site](http://www.mitoproteome.org/data/proteome/index.html)

```{r dataload}
annots.450k <- read.csv(file.path(dataDir, "14_April_2014_Updated_humanmethylation450_15017482_v1-2.csv"), skip=7)
annots.mito <- read.delim(file.path(dataDir, "mitoproteome.tsv"))
```

# Find genomic locations of mitochondrial genes using biomaRt

## setup biomart

```{r setupbiomart}
ensemblmart <-  useMart("ensembl",dataset="hsapiens_gene_ensembl")
attributes <- listAttributes(ensemblmart)
filters <- listFilters(ensemblmart)
```

## pull down chromosomal positions using uniprot ids from mito annotations
- current biomart build is GRCh37
- transcript start and ends
- some ids have multiple transcripts, keep all

```{r getpos}
mito.pos <- getBM(annots.mito$ACCESSION,filters="uniprot_swissprot_accession", attributes=c("uniprot_swissprot_accession","chromosome_name","transcript_start","transcript_end", "strand", "hgnc_symbol"), mart=ensemblmart)
```

# Subset genes to those on nuclear chromosomes  
- dicard all mitochondrial, patches and polymorphic regions
- discard genes without annotated gene symbol
- convert to [BED format](http://genome.ucsc.edu/FAQ/FAQformat.html#format1)

```{r subset}
chrs <- c(seq(1,22), "X", "Y")
mito.pos <- mito.pos[mito.pos$chromosome_name %in% chrs,]

# rearrange columns into bed format
mito.pos$score <- 0
mito.pos <- mito.pos[,c(2,3,4,6,7,5)]
mito.pos$chromosome_name <- paste("chr", mito.pos$chromosome_name, sep="")
# drop genes without known symbol
mito.pos <- mito.pos[which(str_length(mito.pos$hgnc_symbol)>0),]
mito.pos$strand <-   ifelse(mito.pos$strand=="-1", "-", "+")
```

# Mitochondrial genes on nuclear chromosomes

```{r printmitogenes, results='asis'}
kable(mito.pos)
```

# Get coordinates of regions around nuclear mitochondrial genes
3kb upstream of the TSS and 3Kb downstream of transcript end

```{r getpromoters}
mito.pos.flanked <- bedTools.slop(bed1 = mito.pos, opt.string = "-b 3000", genomefile = "~/projects/ab_nucmito/meta/hg19.genome")
mito.pos.upstream <- bedTools.flank(bed1 = mito.pos, opt.string = "-l 3000 -r 0", genomefile = "~/projects/ab_nucmito/meta//hg19.genome")
mito.pos.downstream <- bedTools.flank(bed1 = mito.pos, opt.string = "-l 0 -r 3000", genomefile = "~/projects/ab_nucmito/meta//hg19.genome")
```

# Find 450K probes within flanks and genes of nuclear mitochondrial genes

## Extract coordinates (Human Genome build GRCh37) of 450K probes from annotation file and convert to BED format
```{r 450kextract}
bed.450K <- annots.450k[,c("CHR", "MAPINFO", "Name")]
# adjust to zero-based coordinate format
bed.450K$start <- bed.450K$MAPINFO-1
names(bed.450K) <- c("chr", "end", "name", "start")
bed.450K$chr <- paste("chr", bed.450K$chr, sep="")
bed.450K <- bed.450K[,c("chr", "start", "end", "name")]
```

## Overlap 450K probe coordinates with flanks of nuclear mitochondrial gene coordinates

```{r overlap}
probes450K.in.flanked.nucmitos <- bedTools.2in(bed1=bed.450K,bed2=mito.pos.flanked,opt.string="-wo")
probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos[,c(1,2,3,4,8)]
names(probes450K.in.flanked.nucmitos) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos[with(probes450K.in.flanked.nucmitos, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.flanked.nucmitos <- probes450K.in.flanked.nucmitos[!duplicated(probes450K.in.flanked.nucmitos),]

##########################
##########################
probes450K.in.upstream.nucmitos <- bedTools.2in(bed1=bed.450K,bed2=mito.pos.upstream,opt.string="-wo")
probes450K.in.upstream.nucmitos <- probes450K.in.upstream.nucmitos[,c(1,2,3,4,8)]
names(probes450K.in.upstream.nucmitos) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.upstream.nucmitos <- probes450K.in.upstream.nucmitos[with(probes450K.in.upstream.nucmitos, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.upstream.nucmitos <- probes450K.in.upstream.nucmitos[!duplicated(probes450K.in.upstream.nucmitos),]

##########################
##########################

probes450K.in.downstream.nucmitos <- bedTools.2in(bed1=bed.450K,bed2=mito.pos.downstream,opt.string="-wo")
probes450K.in.downstream.nucmitos <- probes450K.in.downstream.nucmitos[,c(1,2,3,4,8)]
names(probes450K.in.downstream.nucmitos) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.downstream.nucmitos <- probes450K.in.downstream.nucmitos[with(probes450K.in.downstream.nucmitos, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.downstream.nucmitos <- probes450K.in.downstream.nucmitos[!duplicated(probes450K.in.downstream.nucmitos),]

##########################
##########################

probes450K.in.nucmitos <- bedTools.2in(bed1=bed.450K,bed2=mito.pos,opt.string="-wo")
probes450K.in.nucmitos <- probes450K.in.nucmitos[,c(1,2,3,4,8)]
names(probes450K.in.nucmitos) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.nucmitos <- probes450K.in.nucmitos[with(probes450K.in.nucmitos, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.nucmitos <- probes450K.in.nucmitos[!duplicated(probes450K.in.nucmitos),]

```

# Output results

```{r output}
kable(probes450K.in.flanked.nucmitos, row.names=F)
setwd(resultsDir)
write.xlsx2(probes450K.in.flanked.nucmitos, file="probes450K.in.flanked.nucmitos.xlsx")
write.xlsx2(probes450K.in.upstream.nucmitos, file="probes450K.in.upstream.nucmitos.xlsx")
write.xlsx2(probes450K.in.downstream.nucmitos, file="probes450K.in.downstream.nucmitos.xlsx")
write.xlsx2(probes450K.in.nucmitos, file="probes450K.in.nucmitos.xlsx")
```

[Excel file with results](../results/probes450K.in.flanked.nucmitos.xlsx)

# Plot categorical results

```{r plotcategories, dev="svg"}

num.genic <- length(unique(probes450K.in.nucmitos$probeID))
num.upstream <- length(unique(probes450K.in.upstream.nucmitos$probeID))
num.downstream <- length(unique(probes450K.in.downstream.nucmitos$probeID))

nums <- as.data.frame(c(num.upstream,num.genic, num.downstream))
names(nums) <- "count"
nums$category <- c("upstream","genic",  "downstream")
nums <- within(nums, category <- factor(category,levels=c("upstream", "genic","downstream"))) # rearrange factor levels for plot



pdf(file.path(resultsDir, "probe.categories.plot.pdf"))
# plot bars and add text
p <- ggplot(nums, aes(x = category, y = count, fill=factor(category))) +
  geom_bar(stat="identity")+
  scale_fill_manual(values=cbPalette)+
  theme_bw()+ 
  theme(legend.title=element_blank())+
  theme(axis.title.x = element_blank())
p+ geom_text(data=nums,aes(x=category,y=0.5*count,label=count), size=2.5) 
dev.off()


